---
title: "Result_codon_usage_ms"
output:
  word_document: default
  html_document:
    df_print: paged
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

# Codon usage analysis

The reson for perfoming codon usage analysis is to investigate if codon usage bias is presence and if any bias could have an effect on the composition of the coding sequences and if this could influence the estimate of selection and evolutionary rate. 

Four steps
i) Compare codon usage table per branch - between CDS and the aligned geneset
and between branches. 
ii) Compare genewise codon usage frequency (CUF) (codon per 10000) between aligned (4150 genes), most conserved (100 genes), and genes under positive selection in the Leptidea branch
iii) Use branchwise (4150 genes) relative synonymous codon usage (RSCU) and compare with 3rd position GC content. % codons with A/T vs G/C ending with RSCU >1 or <1
iv) Effective codon number (enc), a one dimensional estimate of codon usage from 20 - 61 per gene for each branch. Calc enc* and plot. Model evaluation of variables influencing dnds as a function of branch, enc-exp_enc, GC content and length of alignment. 


```{r codon_usage_data, echo=FALSE}
library(ggplot2)
library(viridis)
library(tidyverse)
library(gridExtra)
library(ggpubr)
library(Hmisc)
library(lme4)
library(car)
library(sjPlot)
library(nord)
library(sjstats)
library(sjPlot)
library(stargazer)
library(rcartocolor)
library(dplyr)


date()
sessionInfo()

#Get the data
cut_branchwise_cds <- read.table("data/cds_CUT_all.table", header = T)
#remove stop codons named *
cut_branchwise_cds <- cut_branchwise_cds[cut_branchwise_cds$AA != "*",]
head(cut_branchwise_cds)
str(cut_branchwise_cds)

cut_branchwise_4150 <- read.table("data/cub_4150_all.csv", header = T)
head(cut_branchwise_4150)
str(cut_branchwise_4150)

#very large wait with reading it in until using
cuf_genewise <- read.table("data/CUT_genewise_frequency_corr.table", header = T)

enc_total <- read.table("data/enc_total.csv", header = T)
#remove the branchspecific namens of the genes, not needed now
enc_total <- enc_total[ ,c(1:3,5,7,9,11,13,15)]
head(enc_total)
str(enc_total)


```


i) Compare codon usage table per branch - between CDS and the aligned geneset
and between branches.


```{r i_test_comp_cds_aligned}

# Wilcoxon rank test for two samples (compare cds and aligned)
wilcox.test(cut_branchwise_cds$Frequency, cut_branchwise_4150$Frequency, paired = T)
#V = 55849, p-value = 0.2217


plot(cut_branchwise_cds$Frequency, cut_branchwise_4150$Frequency)
abline(0,1)
cor.test(cut_branchwise_cds$Frequency, cut_branchwise_4150$Frequency)


cuf_test_df <- rbind(cut_branchwise_cds[,c("Codon", "Branch", "Frequency", "dataset")], cut_branchwise_4150[,c("Codon", "Branch", "Frequency", "dataset")])

kruskal.test(cuf_test_df$Frequency, cuf_test_df$Codon)
#Kruskal-Wallis chi-squared = 897.95, df = 60, p-value < 2.2e-16
kruskal.test(cuf_test_df$Frequency, cuf_test_df$dataset)
#Kruskal-Wallis chi-squared = 0.00091233, df = 1, p-value = 0.9759

ggplot(cuf_test_df, aes(Codon, Frequency)) +
  geom_boxplot(aes(colour = dataset), position = "dodge") +
  stat_compare_means() 

ggplot(cuf_test_df, aes(dataset, Frequency)) +
  geom_boxplot(aes(colour = dataset), position = "dodge") +
  stat_compare_means(method = "t.test", label = "p.signif") +
  facet_grid(~Codon)



```




```{r i_plot, echo=FALSE}
#heatmap

#make one df of both cds and aligned
#add variable with dataset
cut_branchwise_4150$dataset <- as.factor("Aligned")
cut_branchwise_cds$dataset <- as.factor("CDS")

cut_branchwise_comb <- rbind(cut_branchwise_cds[ ,c(1,2,4,6,7)], cut_branchwise_4150[ ,c(1,2,4,6,9)])

#add column with third position
cut_branchwise_comb$third_codon <- as.factor(substr(cut_branchwise_comb$Codon,3,3))
cut_branchwise_comb$third_codon <- factor(x=cut_branchwise_comb$third_codon, levels = c("A","T","G","C"))

#reassign names
levels(cut_branchwise_comb$Branch) <- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae")

#write cuf-table
write.table(as.data.frame(pivot_wider(cut_branchwise_comb, names_from = c("Branch", "dataset"), values_from = "Frequency")), file = "result_files/cuf_cds_aligned.table", sep = " ")


#Heatmap 
ggplot(cut_branchwise_comb) + 
     geom_tile(aes(cut_branchwise_comb$dataset, reorder(cut_branchwise_comb$Codon, as.numeric(cut_branchwise_comb$third_codon)), 
                   group = interaction(cut_branchwise_comb$dataset,cut_branchwise_comb$Branch),
                   fill=cut_branchwise_comb$Frequency), 
               stat = "identity", 
               position = "dodge") + 
     scale_fill_viridis(guide_legend(title = "Codon frequency")) +
     xlab(label = "") +
     ylab(label = "") +
     #scale_x_discrete(breaks = as1:16, labels = group) +
     facet_grid(~cut_branchwise_comb$Branch, switch = "x") +
     theme(panel.background = element_blank(), 
           panel.spacing = unit(0, "lines"),
          strip.background = element_blank(), 
          strip.placement = "outside",
          strip.text = element_text(size = 10, face = "italic"), 
          axis.text = element_text(size = 8, colour = "black"))


```


ii) Compare genewise codon usage frequency (CUF) (codon per 10000) between aligned (4150 genes), most conserved (100 genes), and genes under positive selection in the Leptidea branch

```{r ii_genewise_codon_usage, echo=FALSE}
#remove stop codons
cuf_genewise <- cuf_genewise[,c(1:62)]

result_BS <- read.table("../../../results_summary_200514/summary_results_BS_soft.csv", header = T, row.names = 1)
#filter only 4150 genes
cuf_genewise_filtered <- cuf_genewise[cuf_genewise$GeneID %in% result_BS$GeneID.1, ]

result_BS_100_cons <- read.csv("../../../results_summary_200514/result_BS_100_cons_ordered.csv", header = T, row.names = 1)
#filter only 100 most conserved
cuf_genewise_100_cons <- cuf_genewise[cuf_genewise$GeneID %in% result_BS_100_cons$GeneID.1, ]

result_BS_pos_sel <- read.csv("../../../results_summary_200514/summary_results_BS_sign_soft_filtered.csv", header = T, row.names = 1)
#filter only  pos sel genes
cuf_genewise_pos_sel <- cuf_genewise[cuf_genewise$GeneID %in% result_BS_pos_sel$GeneID.1, ]

#comb df
cuf_means <- rbind(colMeans(cuf_genewise_filtered[,2:62]), colMeans(cuf_genewise_100_cons[,2:62]), colMeans(cuf_genewise_pos_sel[,2:62]))
cuf_means_t <- t(cuf_means)

#difference between pos sel and aligned
cuf_means_diff <- data.frame(cuf_means_t[,3]-cuf_means_t[,1])
cuf_means_diff$Codon <- as.factor(row.names(cuf_means_diff))

#add third codon
cuf_means_diff$third_codon <- as.factor(substr(cuf_means_diff$Codon,3,3))
cuf_means_diff$third_codon <- factor(x=cuf_means_diff$third_codon, levels = c("A","T","G","C"))

#rename colnames(df)[colnames(df) == 'oldName'] <- 'newName'
colnames(cuf_means_diff)[colnames(cuf_means_diff)=="cuf_means_t...3....cuf_means_t...1."] <- "Difference"


#difference between 100 most conserved and aligned
#difference between   pos sel and aligned
cuf_means_diff_100 <- data.frame(cuf_means_t[,2]-cuf_means_t[,1])
cuf_means_diff_100$Codon <- as.factor(row.names(cuf_means_diff_100))

#add third codon
cuf_means_diff_100$third_codon <- as.factor(substr(cuf_means_diff_100$Codon,3,3))
#reorder levels
cuf_means_diff_100$third_codon <- factor(x=cuf_means_diff_100$third_codon, levels = c("A","T","G","C"))

#rename colnames(df)[colnames(df) == 'oldName'] <- 'newName'
colnames(cuf_means_diff_100)[colnames(cuf_means_diff_100)=="cuf_means_t...2....cuf_means_t...1."] <- "Difference"

```


```{r ii_test}
#test pairwise difference of means
kruskal.test(colMeans(cuf_genewise_filtered[,2:62]), colMeans(cuf_genewise_100_cons[,2:62]), colMeans(cuf_genewise_pos_sel[,2:62]))


#test difference as a function of third codon position
anova(lm(cuf_means_diff_100$Difference~cuf_means_diff_100$third_codon))
anova(lm(cuf_means_diff$Difference~cuf_means_diff$third_codon))

#print cuf-table for aligned, 100-cons, pos-sel
write.table(cuf_means_t, file = "result_files/cuf_aligned_100cons_pos_sel.table", col.names = c("Codon", "Aligned", "Conserved", "Pos_sel"))

```

No significat pairwise difference between the different datasets, but grouping the codons after third position gives a difference with higher mean frequencies in G/C-eding codons comp to A/T-ending codons in the cand genes under selection and the nucleotide at the third position can explain a large part of the variance in the difference in mean (anova, F value 30.36, p-value7.37 x 10(-12). No difference in 100 most conserved vs aligned dataset (F value 0.43, p-value 7.36 x 10(-1). 
This suggests that there are no specific codons with bias usage in the cand genes under selection, instead there is a general increase in in usage of codons with G/C ending. In favour of other forces than selection for codon usage that causes the increased GC-content in the cand genes under seletion.

```{r ii_plots, echo=FALSE}
#boxplots, order after third codonposition
par(mfrow=c(3,1))
boxplot(cuf_genewise_filtered[,levels(reorder(cuf_means_diff$Codon, as.numeric(cuf_means_diff$third_codon)))], cex.axis=0.5)
boxplot(cuf_genewise_100_cons[,levels(reorder(cuf_means_diff$Codon, as.numeric(cuf_means_diff$third_codon)))], cex.axis=0.5)
boxplot(cuf_genewise_pos_sel[,levels(reorder(cuf_means_diff$Codon, as.numeric(cuf_means_diff$third_codon)))], cex.axis=0.5)


#Change colour
pos_sel_cuf_plot <- ggplot(cuf_means_diff) +
     geom_hline(yintercept = 0) +
     geom_violin(aes(cuf_means_diff$third_codon, cuf_means_diff$Difference, fill = cuf_means_diff$third_codon)) +
     scale_fill_manual(values = c("#FDE725FF", "#21908CFF", "#440154FF", "grey50", "#3B528BFF"), guide = "none") +
    geom_boxplot(aes(cuf_means_diff$third_codon, cuf_means_diff$Difference, fill = "grey50"), width = 0.1) +
     labs(title = "Candidate genes under positive selection") +
     xlab(label = "Third codon position") +
     ylab(label = "Diff mean frequency") +
     scale_y_continuous(breaks = seq(-8,8,4)) +
     theme(panel.border = element_blank(),
           panel.grid.major.y = element_blank(),
           panel.background = element_blank(),
           axis.line.x.bottom = element_blank(),
           axis.line.y.left = element_line(size = 0.5),
           axis.text = element_text(size = 16), 
           axis.title = element_text(size = 16), 
           axis.ticks.x = element_blank())

pos_sel_cuf_plot

most_conserved_cuf_plot <- ggplot(cuf_means_diff_100) +
     geom_hline(yintercept = 0) +
     geom_violin(aes(cuf_means_diff_100$third_codon, cuf_means_diff_100$Difference, fill = cuf_means_diff_100$third_codon)) +
     scale_fill_manual(values = c("#FDE725FF", "#21908CFF", "#440154FF", "grey50", "#3B528BFF"), guide = "none") +
    geom_boxplot(aes(cuf_means_diff_100$third_codon, cuf_means_diff_100$Difference, fill = "grey50"), width = 0.1) +
     labs(title = "100 most conserved genes") +
     xlab(label = "") +
     ylab(label = "Diff mean frequency") +
     scale_y_continuous(breaks = seq(-8,8,4)) +
     theme(panel.border = element_blank(),
           panel.grid.major.y = element_blank(),
           panel.background = element_blank(),
           axis.line.x.bottom = element_blank(),
           axis.line.y.left = element_line(size = 0.5),
           axis.text = element_text(size = 16), 
           axis.title = element_text(size = 16), 
           axis.ticks.x = element_blank())

grid.arrange(most_conserved_cuf_plot, pos_sel_cuf_plot)

#test colours
pos_sel_cuf_plot <- ggplot(cuf_means_diff) +
     geom_hline(yintercept = 0, size = 0.2) +
     geom_violin(aes(cuf_means_diff$third_codon, cuf_means_diff$Difference, fill = cuf_means_diff$third_codon)) +
     scale_fill_carto_d(type = "qualitative",
                        palette = "Earth", 
                       guide = "none") +
    geom_boxplot(aes(cuf_means_diff$third_codon, cuf_means_diff$Difference), fill = "grey50", width = 0.1) +
     #labs(title = "Candidate genes under positive selection") +
     xlab(label = "Third codon position") +
     ylab(label = "Diff mean frequency") +
     scale_y_continuous(breaks = seq(-8,8,4)) +
     theme(panel.border = element_blank(),
           panel.grid.major.y = element_blank(),
           panel.background = element_blank(),
           axis.line.x.bottom = element_blank(),
           axis.line.y.left = element_line(size = 0.2),
           axis.text = element_text(size = 10, colour = "black"), 
           axis.title = element_text(size = 10, colour = "black"), 
           axis.ticks.x = element_blank(),
           title = element_text(size = 10))



most_conserved_cuf_plot <- ggplot(cuf_means_diff_100) +
     geom_hline(yintercept = 0, size = 0.2) +
     geom_violin(aes(cuf_means_diff_100$third_codon, cuf_means_diff_100$Difference, fill = cuf_means_diff_100$third_codon)) +
     scale_fill_carto_d(type = "qualitative",
                        palette = "Earth",
                       guide = "none") +
     geom_boxplot(aes(cuf_means_diff_100$third_codon, cuf_means_diff_100$Difference), fill = "grey50", width = 0.1) +
     #labs(title = "100 most conserved genes") +
     xlab(label = "") +
     ylab(label = "Diff mean frequency") +
     scale_y_continuous(breaks = seq(-8,8,4)) +
     theme(panel.border = element_blank(),
           panel.grid.major.y = element_blank(),
           panel.background = element_blank(),
           axis.line.x.bottom = element_blank(),
           axis.line.y.left = element_line(size = 0.2),
           axis.text = element_text(size = 10), 
           axis.title = element_text(size = 10), 
           axis.ticks.x = element_blank(), 
           title = element_text(size = 10))


diff_plot <- ggarrange(most_conserved_cuf_plot, pos_sel_cuf_plot, nrow = 2, ncol = 1)

ggsave("figures/diff_cons_pos_sel.pdf", plot = diff_plot, device = "pdf", height = 6, width = 8)
ggsave("figures/diff_cons_pos_sel.jpg", plot = diff_plot, device = "jpg", height = 6, width = 8)
```


iii) Codon usage bias. Investigate the synonymous codon usage and compare to the GC content in third position, branchwise.

```{r RSCU_branchwise_cds, echo=FALSE}

cut_branchwise_cds$third_codon <- as.factor(substr(cut_branchwise_cds$Codon,3,3))

#reorder levels
cut_branchwise_cds$third_codon <- factor(x=cut_branchwise_cds$third_codon, levels = c("A","T","G","C"))

cut_branchwise_cds$Null_freq <- cut_branchwise_4150$Null_freq

cut_branchwise_cds$RSCU <- cut_branchwise_cds$Fraction/cut_branchwise_cds$Null_freq

#Get a table of codon usage bias by third position codon and branch
RSCU_summary_branch <- aggregate(cut_branchwise_cds[cut_branchwise_cds$RSCU > 1.0,"third_codon"], list(cut_branchwise_cds[cut_branchwise_cds$RSCU > 1.0,"Branch"]), summary)

RSCU_summary_branch_table_cds <- rbind(as.data.frame(RSCU_summary_branch[ ,2]), colSums(as.data.frame(RSCU_summary_branch[ ,2])), summary(cut_branchwise_cds$third_codon))

rownames(RSCU_summary_branch_table)<- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae", "Nr codons RSCU>1", "Total nr of codons")

write.table(spread(cut_branchwise_cds[,c(1,2,6,8,9)], "Branch", "RSCU"), file = "result_files/RSCU_cds.table", col.names = c("Codon", "AA", "Third codon", "B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae"))

write.table(RSCU_summary_branch_table_cds, file = "result_files/RSCU_ending_cds.table")
```


```{r RSCU_branchwise_4150, echo=FALSE}
cut_branchwise_4150_RSCU <- cut_branchwise_4150[ ,c(1,2,6,8)]
#remove non redundant codons
cut_branchwise_4150_RSCU <- cut_branchwise_4150_RSCU[cut_branchwise_4150_RSCU$Codon != "TGG" & cut_branchwise_4150_RSCU$Codon != "ATG", ]
#add column with third position
cut_branchwise_4150_RSCU$third_codon <- as.factor(substr(cut_branchwise_4150_RSCU$Codon,3,3))

#reorder levels
cut_branchwise_4150_RSCU$third_codon <- factor(x=cut_branchwise_4150_RSCU$third_codon, levels = c("A","T","G","C"))

#Get a table of codon usage bias by third position codon and branch
RSCU_summary_branch <- aggregate(cut_branchwise_4150_RSCU[cut_branchwise_4150_RSCU$RSCU > 1.0,5], list(cut_branchwise_4150_RSCU[cut_branchwise_4150_RSCU$RSCU > 1.0,3]), summary)

RSCU_summary_branch_table <- rbind(as.data.frame(RSCU_summary_branch[ ,2]), colSums(as.data.frame(RSCU_summary_branch[ ,2])), summary(cut_branchwise_4150_RSCU$third_codon))

rownames(RSCU_summary_branch_table)<- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae", "Nr codons RSCU>1", "Total nr of codons")

write.table(spread(cut_branchwise_4150_RSCU, "Branch", "RSCU"), file = "result_files/RSCU_aligned.table", col.names = c("Codon", "AA", "Third codon", "B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae"))

write.table(RSCU_summary_branch_table, file = "result_files/RSCU_ending.table")

```

```{r iii_test}
#test RSCU as a function of 
capture.output(anova(lm(cut_branchwise_4150_RSCU$RSCU~cut_branchwise_4150_RSCU$Branch + cut_branchwise_4150_RSCU$third_codon + cut_branchwise_4150_RSCU$Branch:cut_branchwise_4150_RSCU$third_codon)), file = "result_files/anova_RSCU.txt")

mean(cut_branchwise_4150_RSCU)

ggplot(cut_branchwise_4150_RSCU) +
  boxplot()

```

Most of the variance in RSCU can be explained by the third codon position (anova, F 25.01, p-value 4.82 x 10 (-15)), and an interaction between branch and third codon position(anova, ). This difference is mostly due to increased use of A/T ending coons compared to G/C-ending codons, with the exeption of C. cecrops that have more G/C-ending codons with RSCU above 1.


```{r iii_plot, echo=FALSE}

RSCU_summary_branch_table


#reorder
cut_branchwise_4150_RSCU$Branch <- factor(cut_branchwise_4150_RSCU$Branch, levels=c("b_mori", "p_mac", "l_acc","l_sin","p_sen","c_cecr","d_plex", "h_mel"))


#heatmap RSCU old
ggplot(cut_branchwise_4150_RSCU) +
    geom_tile(aes(cut_branchwise_4150_RSCU$Branch, reorder(cut_branchwise_4150_RSCU$Codon, as.numeric(cut_branchwise_4150_RSCU$third_codon)),
                  fill=cut_branchwise_4150_RSCU$RSCU)) +
    scale_fill_gradient2(midpoint = 1, low = "blue", high = "darkgreen", guide_legend(title = "RSCU")) +
    xlab(label = "") +
    ylab(label = "") +
    scale_x_discrete(labels = c("l_acc" = "L. accius", "p_sen" = "P. sennae", "l_sin" = "L. sinapis", "c_cecr" = "C. cecrops", "d_plex" = "D. plexippus", "h_mel" = "H. melpomene", "p_mac" = "P. machaon", "b_mori" = "B. mori")) +
    theme(panel.background = element_blank(),
          axis.text.x = element_text(size = 10, colour = "black", face = "italic"),
          axis.text.y = element_text(size = 8))

#ms
ggplot(cut_branchwise_4150_RSCU) + 
    geom_tile(aes(cut_branchwise_4150_RSCU$Branch, reorder(cut_branchwise_4150_RSCU$Codon, as.numeric(cut_branchwise_4150_RSCU$third_codon)), 
                  fill=cut_branchwise_4150_RSCU$RSCU)) + 
    scale_fill_gradient2(midpoint = 1, low = "blue", high = "darkgreen", guide_legend(title = "RSCU")) +
    xlab(label = "") +
    ylab(label = "") +
    scale_x_discrete(limits=c("b_mori", "p_mac", "l_acc","l_sin","p_sen","c_cecr","d_plex", "h_mel"),
      labels = c("l_acc" = "L. accius", "p_sen" = "P. sennae", "l_sin" = "L. sinapis", "c_cecr" = "C. cecrops", "d_plex" = "D. plexippus", "h_mel" = "H. melpomene", "p_mac" = "P. machaon", "b_mori" = "B. mori")) +
    theme(panel.background = element_blank(), 
          axis.text.x = element_text(size = 10, colour = "black", face = "italic", angle = 45, hjust = 1),
          axis.text.y = element_text(size = 8),
          axis.ticks.x = element_line(size = 0.2),
          axis.ticks.y = element_blank())

ggsave("figures/heatmap_RSCU.pdf", device = "pdf", height = 6, width = 6)

#new colour
ggplot(cut_branchwise_4150_RSCU) + 
    geom_tile(aes(cut_branchwise_4150_RSCU$Branch, reorder(cut_branchwise_4150_RSCU$Codon, as.numeric(cut_branchwise_4150_RSCU$third_codon)), 
                  fill=cut_branchwise_4150_RSCU$RSCU)) + 
    scale_fill_carto_c(palette = "Geyser", type = "diverging", guide_legend(title = "RSCU"), limits = c(0,2)) +
    xlab(label = "") +
    ylab(label = "") +
    scale_x_discrete(limits=c("b_mori", "p_mac", "l_acc","l_sin","p_sen","c_cecr","d_plex", "h_mel"),
      labels = c("l_acc" = "L. accius", "p_sen" = "P. sennae", "l_sin" = "L. sinapis", "c_cecr" = "C. cecrops", "d_plex" = "D. plexippus", "h_mel" = "H. melpomene", "p_mac" = "P. machaon", "b_mori" = "B. mori")) +
    theme(panel.background = element_blank(), 
          axis.text.x = element_text(size = 10, colour = "black", face = "italic", angle = 45, hjust = 1),
          axis.text.y = element_text(size = 8),
          axis.ticks.x = element_line(size = 0.2),
          axis.ticks.y = element_blank())

ggsave("figures/heatmap_RSCU_geyser.pdf", device = "pdf", height = 6, width = 6)


#violin plot
ggplot(cut_branchwise_4150_RSCU) +
  geom_hline(yintercept = 1.0, lty = 2, colour = "grey") +
  geom_violin(aes(cut_branchwise_4150_RSCU$third_codon, cut_branchwise_4150_RSCU$RSCU, group=interaction(cut_branchwise_4150_RSCU$third_codon, cut_branchwise_4150_RSCU$Branch), fill = cut_branchwise_4150_RSCU$Branch)) +
  scale_fill_viridis(discrete = T,
                     labels = c("l_acc" = "L. accius", "p_sen" = "P. sennae", "l_sin" = "L. sinapis", "c_cecr" = "C. cecrops", "d_plex" = "D. plexippus", "h_mel" = "H. melpomene", "p_mac" = "P. machaon", "b_mori" = "B. mori"),
                     guide_legend(title = "Branch")) +
  ylab(label = "RSCU") +
  xlab(label = "Third codon position") +
  theme(panel.background = element_blank(), 
        panel.grid.major.y = element_blank(),
        axis.line.y.left = element_line(size = 0.5),
        axis.line.x.bottom = element_line(size = 0.5),
        axis.text = element_text(size = 16), 
        axis.title = element_text(size = 16) 
        )

branch_colours <- nord(n = 8, palette = "red_mountain")

ggplot(cut_branchwise_4150_RSCU) +
  geom_hline(yintercept = 1.0, lty = 2, colour = "grey") +
  geom_violin(aes(cut_branchwise_4150_RSCU$third_codon, cut_branchwise_4150_RSCU$RSCU, group=interaction(cut_branchwise_4150_RSCU$third_codon, cut_branchwise_4150_RSCU$Branch), fill = cut_branchwise_4150_RSCU$Branch)) +
  scale_fill_manual(values = branch_colours,
                    labels = c("l_acc" = "L. accius", "p_sen" = "P. sennae", "l_sin" = "L. sinapis", "c_cecr" = "C. cecrops", "d_plex" = "D. plexippus", "h_mel" = "H. melpomene", "p_mac" = "P. machaon", "b_mori" = "B. mori"),
                     guide_legend(title = "Branch")) +
  ylab(label = "RSCU") +
  xlab(label = "Third codon position") +
  theme(panel.background = element_blank(), 
        panel.grid.major.y = element_blank(),
        axis.line.y.left = element_line(size = 0.5),
        axis.line.x.bottom = element_line(size = 0.5),
        axis.text = element_text(size = 16), 
        axis.title = element_text(size = 16) 
        )

ggsave("figures/violin_RSCU.pdf", device = "pdf", height = 6, width = 12)

```




iv) Effective codon number (enc), a one dimensional estimate of codon usage from 20 - 61 per gene for each branch. Calc enc* and plot. Calc enc_diff.  Model evaluation of variables influencing dnds as a function of branch, enc, enc-exp_enc, GC content and length of alignment. 

```{r iv_enc, echo=FALSE}
#filter to 4150 genes
enc_total <- read.table("data/enc_total.csv", header = T)
#remove the branchspecific namens of the genes, not needed now
enc_total <- enc_total[ ,c(1:3,5,7,9,11,13,15)]
head(enc_total)
str(enc_total)

result_BS <- read.table("../../../results_summary_200514/summary_results_BS_soft.csv", header = T, row.names = 1)

enc_total_4150 <- enc_total[enc_total$GeneID %in% result_BS$GeneID.1, ]


##############################

#add dataset
result_BS_100_cons_ordered <- read.csv( "../../../results_summary_200514/result_BS_100_cons_ordered.csv")
result_BS_100_cons_ordered$dataset <- "cons"

#dataset <- rbind(result_BS_sign_filtered[, c("GeneID.1","dataset")], result_BS_100_cons_ordered[, c("GeneID.1","dataset")])
                 
result_BS$dataset <- "neutral"
rest_dataset <- anti_join(result_BS[,c("GeneID.1", "dataset")], result_BS_100_cons_ordered[, c("GeneID.1","dataset")], by = "GeneID.1")
rest_dataset$GeneID.1 <- as.character(rest_dataset$GeneID.1)

dataset <- rbind(result_BS_100_cons_ordered[, c("GeneID.1","dataset")], rest_dataset[, c("GeneID.1","dataset")])
colnames(dataset) <- c("GeneID", "dataset")

enc_total_4150 <- left_join(enc_total_4150, dataset, by = "GeneID")

############################
#add gc
enc_total_4150_long <- gather(enc_total_4150, branch, enc, 2:9)
enc_total_4150_long$branch <- as.factor(enc_total_4150_long$branch)
#rename levels
levels(enc_total_4150_long$branch) <- c("B_mori", "C_cecrops", "D_plexippus", "H_melpomene", "L_accius", "L_sinapis", "P_machaon", "P_sennae")

gc_content_unfiltered <- read.table("../../../results_summary_200514/Summary_gc_total_positions.csv", header = T)

gc_content_4150 <- gc_content_unfiltered[gc_content_unfiltered$GeneID %in% result_BS$GeneID.1, ]

gc_content_4150_long <- gather(gc_content_4150[ ,c(1,2,4:11)], branch, gc, 3:10)
gc_4150_positions <- spread(gc_content_4150_long, position, gc)
names(gc_4150_positions)[names(gc_4150_positions) == "1"] <- "gc_1"
names(gc_4150_positions)[names(gc_4150_positions) == "2"] <- "gc_2"
names(gc_4150_positions)[names(gc_4150_positions) == "3"] <- "gc_3"
gc_4150_positions$branch <- as.factor(gc_4150_positions$branch)


#merge gc and enc
enc_gc_4150 <- merge(gc_4150_positions, enc_total_4150_long, by = c("GeneID", "branch"))


#get third position
enc_gc3 <- gc_content_4150[gc_content_4150$position == 3,]

#rename column
#add alignment length
enc_gc3$al_length <- result_BS$al_length

#reformat to long
enc_gc_long1 <- gather(enc_gc3[,c(1,4:12)], branch, gc3, 2:9)

#merge with enc_gc, naming it enc_gc_long to make it easier downstream
enc_gc_long <- merge(enc_gc_4150, enc_gc_long1)
enc_gc_long$gc3 <- NULL



```

```{r iv_test_branch_enc}
#check branches

boxplot(enc_total_4150[,2:9])

hist(enc_gc_long$enc, breaks = 500)

mod_norm <- lmer(enc~branch + (1|GeneID), data = enc_gc_long)
mod_norm_inv <- glmer(enc~branch + (1|GeneID), data = enc_gc_long, family = inverse.gaussian(link = "1/mu^2"))
mod_inv <- glmer(enc~branch + (1|GeneID), data = enc_gc_long, family = Gamma(link = "inverse"))
mod_log <- glmer(enc~branch + (1|GeneID), data = enc_gc_long, family = Gamma(link = "log"))

tab_model(mod_norm_inv)
plot(mod_norm)
#resdiual do not look ok... dependent variable is capped at 61...

#use a non-parametric test
kruskal.test(enc_gc_long$enc, enc_gc_long$branch)
#Kruskal-Wallis chi-squared = 669.03, df = 7, p-value < 2.2e-16


```


```{r iv_exp_enc, echo=FALSE}
#exp ENC
#ENC=2+GC3+29/(GC3^2+(1-GC)^2)
enc_gc_long$exp_enc <- (2+enc_gc_long$gc_3+29/(enc_gc_long$gc_3^2+(1-enc_gc_long$gc_3)^2))

#difference between observed and expected, normalised by exp value (as proportion of exp_enc)
enc_gc_long$diff_enc <- (enc_gc_long$exp_enc-enc_gc_long$enc)/enc_gc_long$exp_enc


#add substitution per type
subst_type_gccons_tot <- 
read.table(file = "../08_subst_type/08_03_subst_type_R/data/subst_type_gccons_tot_gccontent.table", header = T)


#change gene-id colname
subst_type_gccons_tot$GeneID <- subst_type_gccons_tot$gene_ID
#combine with enc
enc_gc_type <- 
join(subst_type_gccons_tot[,-1], enc_gc_long) 

enc_gc_type$type <- as.factor(enc_gc_type$type)
enc_gc_type$branch <- as.factor(enc_gc_type$branch)
levels(enc_gc_type$branch)
levels(enc_gc_type$branch) <- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae")

#reorder
enc_gc_type$branch <- factor(enc_gc_type$branch, levels=c("B. mori","P. machaon", "L. accius", "L. sinapis", "P. sennae", "C. cecrops", "D. plexippus", "H. melpomene"))

#enc_exp according to dos Reis et al 2004
#Nc_exp=a + GC3 + b/(gc3^2 + (c-GC3)^2)

enc_gc_type$exp_enc_dR <- (-6+enc_gc_type$gc_3+34/(enc_gc_type$gc_3^2+(1.025-enc_gc_type$gc_3)^2))
enc_gc_type$diff_enc_dR <- (enc_gc_type$exp_enc_dR-enc_gc_type$enc)/enc_gc_type$exp_enc_dR

plot(enc_gc_type$diff_enc_dR, enc_gc_type$diff_enc)

write.table(enc_gc_type, "~/Documents/GitHub/Project01_comp_gen/scripts_200505/07_codon_usage/07_03_codon_usage_analysis_r/result_files/enc_gc_types_dR.table")

enc_gc_type <- read.table("~/Documents/GitHub/Project01_comp_gen/scripts_200505/07_codon_usage/07_03_codon_usage_analysis_r/result_files/enc_gc_types_dR.table", header = T)



length(unique(enc_gc_type$GeneID)) 

```

```{r iv_plot_ds_types}
#get data if not in environment
subst_type_gccons_tot <- 
read.table(file = "../../08_subst_type/08_03_subst_type_R/data/subst_type_gccons_tot_gccontent.table", header = T)

subst_type_gccons_tot$branch <- as.factor(subst_type_gccons_tot$branch)
#change order
levels(subst_type_gccons_tot$branch)
#[1] "B_mori"      "C_cecrops"   "D_plexippus" "H_melpomene" "L_accius"    "L_sinapis"   "P_machaon"   "P_sennae"   
#reassign
levels(subst_type_gccons_tot$branch) <- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae")

#reorder
subst_type_gccons_tot$branch <- factor(subst_type_gccons_tot$branch, levels=c("B. mori","P. machaon", "L. accius", "L. sinapis", "P. sennae", "C. cecrops", "D. plexippus", "H. melpomene"))

subst_type_gccons_tot$GeneID <- subst_type_gccons_tot$gene_ID
#combine with enc
enc_gc_type <- 
join(subst_type_gccons_tot[,-1], enc_dnds_ds_dn[ ,c("GeneID", "branch", "enc", "exp_enc", "diff_enc", "al_length", "dataset")]) 

enc_gc_type$type <- as.factor(enc_gc_type$type)

enc_gc_type %>%
  filter(type %in% "gc_cons")
```

```{r iv_exp_enc_test}
######
#old
#anova test to check variables, dnds as a function of the difference in enc-exp_enc
summary(aov(enc_dnds_3280$dnds~(enc_dnds_3280$diff_enc + enc_dnds_3280$enc + enc_dnds_3280$Branch + enc_dnds_3280$gc + enc_dnds_3280$al_length + enc_dnds_3280$diff_enc:enc_dnds_3280$Branch)))

summary(aov(enc_dnds_ds$dS~(enc_dnds_ds$diff_enc + enc_dnds_ds$enc + enc_dnds_ds$Branch + enc_dnds_ds$gc + enc_dnds_ds$diff_enc:enc_dnds_ds$Branch)))

summary(aov(enc_dnds_ds$dN~(enc_dnds_ds$diff_enc + enc_dnds_ds$enc + enc_dnds_ds$Branch + enc_dnds_ds$gc + enc_dnds_ds$diff_enc:enc_dnds_ds$Branch)))
#########################

#read data if not in environment
enc_dnds_3280 <- read.table(file = "../../07_codon_usage/07_03_codon_usage_analysis_r/data/enc_dnds_3280.table", sep = "")


#evaluating normality of the responsvariable
hist(enc_dnds_3280$dnds, breaks = 1000)
hist(log(enc_dnds_3280$dnds), breaks = 1000)
#log could approximate normality

#scale the explanatory variables
enc_dnds_3280_scaled <- scale(enc_dnds_3280[ ,4:10], scale = T, center = T)
enc_dnds_3280_scaled <- cbind(enc_dnds_3280[,1:3], enc_dnds_3280_scaled)

#random factor that we want to control for are GeneID, branch
#fixed factors are the rest
enc_1.lmer <- lmer(log(dnds) ~ enc + gc_1 + gc_2 + gc_3 + diff_enc + al_length + (1|GeneID) + (1|branch) , data = enc_dnds_3280_scaled)
summary(enc_1.lmer)

capture.output(summary(enc_1.lmer), file = "../../07_codon_usage/07_03_codon_usage_analysis_r/result_files/model_enc_gc_dnds.txt")





#write mean enc, enc_exp and enc diff to table 
enc_colmean_sd <- cbind(as.data.frame(colMeans(enc_total_4150[2:9])), as.data.frame(apply(enc_total_4150[2:9], 2, sd)), as.data.frame(colMeans(spread(enc_gc_long[,c(1,3,6)], "Branch", "exp_enc")[2:9])), as.data.frame(apply(spread(enc_gc_long[,c(1,3,6)], "Branch", "exp_enc")[2:9], 2, sd)), as.data.frame(colMeans(spread(enc_gc_long[,c(1,3,7)], "Branch", "diff_enc")[2:9])), as.data.frame(apply(spread(enc_gc_long[,c(1,3,7)], "Branch", "diff_enc")[2:9], 2, sd)))

write.table(enc_colmean_sd, file = "result_files/enc_mean.table", col.names = c("enc_mean", "enc_sd", "exp_enc_mean", "exp_enc_sd", "diff_enc_mean", "diff_enc_sd"), row.names = c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae"))

#test corr enc gc
capture.output(cor.test(((enc_gc_long$gc_1 + enc_gc_long$gc_2)/2), enc_gc_long$enc, method = "spearman"), file = "result_files/corr_enc_gc.txt")

capture.output(cor.test(((enc_gc_long$gc_1 + enc_gc_long$gc_2 + enc_gc_long$gc_3)/3), enc_gc_long$enc, method = "spearman"), file = "result_files/corr_enc_gc.txt", append = T)
#test corr diff enc and gc
capture.output(cor.test(((enc_gc_long$gc_1 + enc_gc_long$gc_2)/2), enc_gc_long$diff_enc, method = "spearman"), file = "result_files/corr_enc_gc.txt", append = T)

capture.output(cor.test(((enc_gc_long$gc_1 + enc_gc_long$gc_2 + enc_gc_long$gc_3)/3), enc_gc_long$diff_enc, method = "spearman"), file = "result_files/corr_enc_gc.txt", append = T)
enc_gc_long$gc_tot <- (enc_gc_long$gc_1 + enc_gc_long$gc_2 + enc_gc_long$gc_3)/3
summary(lm(enc~gc_tot, data=enc_gc_long))


#corr enc vs synonymous subst


```

The difference between expected enc and the observed enc could be interpreted as the strength of codon usage bias. This difference does not explain the variance in dnds, the variance in dnds is best explained by branch and third position gc content.

```{r iv_plot_gc, echo=FALSE}
#enc_dnds_3280 <- read.table(file = "../../07_codon_usage/07_03_codon_usage_analysis_r/data/enc_dnds_3280.table")

#reassign
#levels(enc_gc_long$branch) <- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae")

#reorder
#enc_gc_long$branch <- factor(enc_gc_long$branch, levels=c("B. mori","P. machaon", "L. accius", "L. sinapis", "P. sennae", "C. cecrops", "D. plexippus", "H. melpomene"))

#write.table(enc_gc_long, file = "data/enc_gc_long.table")

enc_gc_long <- read.table(file = "data/enc_gc_long.table", header = T)

enc_gc_long$branch <- factor(enc_gc_long$branch, levels=c("B. mori","P. machaon", "L. accius", "L. sinapis", "P. sennae", "C. cecrops", "D. plexippus", "H. melpomene"))

gc_colours <- nord(n=3, "red_mountain")
# "#7D3232" "#D4AF7D" "#32324B"

#gc1 and2 vs cg3 (neutrality plot?)
plot_corr_gc12_gc3 <- ggplot(enc_gc_long, aes(((gc_1 + gc_2)/2)*100, gc_3*100)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y = 100, size = 2.5) +
  scale_y_continuous(name = "GC-content third position (%)", 
                     limits = c(10, 100)) +
  scale_x_continuous(name = "GC-content first and second position (%)") +
  scale_fill_gradient(high = "#32324B", low = "light grey") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))


ggsave(plot_corr_gc12_gc3, filename = "figures/plot_corr_gc12_gc3.pdf", device = "pdf", height = 6, width = 8)


#corr enc and gc 1st and 2nd pos 
plot_corr_enc_gc <- ggplot(enc_gc_long, aes(((gc_1 + gc_2)/2)*100, enc)) +
  geom_bin2d() +
  geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "centre", label.y = 65) +
  scale_y_continuous(name = "Observed ENC") +
  scale_x_continuous(name = "GC-content 1st and 2nd position") +
  scale_fill_gradient(high = "white", low = "#7D3232") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_gc, file="figures/corr_enc_gc.pdf", device = "pdf",height = 6, width=8)


plot_corr_enc_gctot <- ggplot(enc_gc_long, aes(((gc_1 + gc_2 + gc_3)/3)*100, enc)) +
  geom_bin2d() +
  geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "centre", label.y = 65) +
  scale_y_continuous(name = "Observed ENC") +
  scale_x_continuous(name = "GC-content") +
  scale_fill_gradient(high = "white", low = "#7D3232") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_gctot, file="figures/corr_enc_gc_tot.pdf", device = "pdf",height = 6, width=8)

#incl exp_enc
plot_corr_enc_exp_enc_gc <- ggplot(enc_gc_long, aes(((gc_1 + gc_2 + gc_3)/3)*100, enc)) +
  geom_bin2d() +
  geom_smooth(mapping = aes(((gc_1 + gc_2 + gc_3)/3)*100, exp_enc), 
              colour ="#494446",
              size = 1.5, 
              linetype="dashed", method="loess", se=FALSE) +
  geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "centre", label.y = 65) +
  scale_y_continuous(name = "Observed ENC") +
  scale_x_continuous(name = "GC-content") +
  scale_fill_gradient(high = "white", low = "#7D3232") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_exp_enc_gc, file="figures/corr_enc_exp_enc_gc.pdf", device = "pdf",height = 6, width=8)


plot_corr_enc_gc_diff <- ggplot(enc_gc_long, aes(((gc_1 + gc_2 + gc_3)/3)*100, diff_enc)) +
  geom_bin2d() +
  geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "centre", label.y = 0.40) +
  scale_y_continuous(name = "ENC_diff") +
  scale_x_continuous(name = "GC-content") +
  scale_fill_gradient(high = "white", low = "#7D3232") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_gc_diff_v2, file="figures/corr_enc_gc_diff.pdf", device = "pdf",height = 6, width=8)
#fae6de, #541608
plot_corr_enc_gc_diff_v2 <- ggplot(enc_gc_long, aes(((gc_1 + gc_2 + gc_3)/3)*100, diff_enc)) +
  geom_bin2d() +
  geom_smooth(method = "lm", colour="black", size = 0.4) +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "centre", label.y = 0.50) +
  scale_y_continuous(name = "ENC_diff") +
  scale_x_continuous(name = "GC-content") +
  scale_fill_gradient(high = "#541608", low = "#d6b8ab") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))


plot_corr_enc_diff_gc3 <- ggplot(enc_gc_long, aes(gc_3*100, diff_enc)) +
  geom_bin2d() +
  geom_smooth(method = "lm", colour="black", size = 0.4) +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y = -0.40) +
  scale_y_continuous(name = "ENC_diff") +
  scale_x_continuous(name = "GC-content in third position (%)") +
  scale_fill_gradient(high = "#541608", low = "#d6b8ab") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_diff_gc3, file="figures/corr_enc_diff_gc3.pdf", device = "pdf",height = 6, width=8)

#enc obs
plot_corr_enc_gc3 <- ggplot(enc_gc_long, aes(gc_3*100, enc)) +
  geom_bin2d() +
  geom_smooth(method = "lm", colour="black", size = 0.4) +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y = 10) +
  scale_y_continuous(name = expression("ENC"[obs])) +
  scale_x_continuous(name = "GC-content in third position (%)") +
  scale_fill_gradient(high = "#541608", low = "#d6b8ab") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_gc3, file="figures/corr_enc_gc3.pdf", device = "pdf",height = 6, width=8)

#in one plot
plot_corr_enc_diff_gc3_all <- ggplot(enc_gc_long, aes(gc_3*100, diff_enc)) +
  geom_point(aes(colour = branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm") +
  #stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y = -0.40) +
  scale_y_continuous(name = expression("ENC"[diff])) +
  scale_x_continuous(name = "GC-content in third position (%)") +
  scale_color_manual(values = branch_colours) +
  #scale_fill_gradient(high = "#541608", low = "#d6b8ab") +
  #facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_line(size = 0.2, colour = "black"),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"))

#enc obs
plot_corr_enc_gc3_all <- ggplot(enc_gc_long, aes(gc_3*100, enc)) +
  geom_point(aes(colour = branch), alpha=0.05) +
  #geom_smooth(aes(colour = branch), method = "lm") +
  geom_smooth(aes(colour = branch), method = "loess") +
  geom_smooth(aes(gc_3*100, exp_enc), colour = "#494446", linetype = "dashed", method = "loess") +
  #stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y = -0.40) +
  scale_y_continuous(name = expression("ENC"[obs])) +
  scale_x_continuous(name = "GC-content in third position (%)") +
  scale_color_manual(values = branch_colours) +
  #scale_fill_gradient(high = "#541608", low = "#d6b8ab") +
  #facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_line(size = 0.2, colour = "black"),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"))

#ggsave(plot=
#         ggarrange(plot_corr_enc_gc3_all, plot_corr_enc_diff_gc3_all, common.legend = T), 
#                   file="figures/corr_enc_gc3.pdf", device = "pdf",height = 6, width=8)

#dnds_enc <- 
  ggplot(enc_dnds_ds_dn, aes(enc, dnds)) +
  geom_point(aes(colour = branch), alpha = 0.05) +
  geom_smooth(aes(colour = branch), method="loess") +
  geom_smooth(aes(colour = branch), method = "lm") +
  #stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10", 
                     name = expression("log("*omega*")")) +
  scale_x_continuous(name = expression("ENC"[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"))

enc_dnds_ds_dn %>%
  ggplot(aes(enc, dnds)) +
  geom_point(aes(colour = branch), alpha = 0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  #stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10",
                     limits = c(0.001,1),
                     name = expression(omega)) +
  scale_x_continuous(name = expression("ENC"[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"))

#binned
enc_dnds_ds_dn %>%
  mutate(dnds_bin = cut(dnds, breaks=10)) %>%
  ggplot(aes(dnds_bin, enc)) +
  geom_boxplot() +
  scale_x_discrete(name = expression(omega)) +
  scale_y_continuous(name = expression("ENC"[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"))
  

enc_dnds_ds_dn %>%
  mutate(enc_bin = cut(enc, breaks=c(10, 20, 30, 40, 50, 61))) %>%
  ggplot(aes(enc_bin, dnds)) +
  geom_boxplot() +
  scale_x_discrete(name = expression("ENC"[obs])) +
  scale_y_continuous(trans = "log10", 
                     name = expression("log("*omega*")")) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"))


#bin in quantiles
enc_dnds_ds_dn %>%
  mutate(diff_bin = cut(diff_enc, breaks=c(floor(min(diff_enc)), round(quantile(diff_enc)[[2]],3), round(quantile(diff_enc)[[3]],3), round(quantile(diff_enc)[[4]],3), ceiling(max(diff_enc))))) %>%
  ggplot(aes(diff_bin, dnds, fill=branch)) +
  geom_boxplot() +
  geom_hline(yintercept = median(enc_dnds_ds_dn$dnds)) +
  #geom_point() +
  stat_compare_means() +
  scale_x_discrete(name = expression("ENC"[diff])) +
  scale_y_continuous(trans = "log10", 
                     name = expression(omega)) +
  scale_fill_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black")) #+
  facet_grid(~branch)

enc_dnds_ds_dn %>%
  mutate(diff_bin = cut(diff_enc, breaks=c(floor(min(diff_enc)), round(quantile(diff_enc)[[2]],3), round(quantile(diff_enc)[[3]],3), round(quantile(diff_enc)[[4]],3), ceiling(max(diff_enc))))) %>%
  ggplot(aes(branch, dnds, fill=diff_bin)) +
  geom_boxplot() +
  geom_hline(yintercept = median(enc_dnds_ds_dn$dnds)) +
  #geom_point() +
  stat_compare_means(label.sep = "\n") +
  scale_x_discrete(name = "") +
  scale_y_continuous(trans = "log10", 
                     name = expression(omega)) +
  scale_fill_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black")) #+
  facet_grid(~branch)

enc_dnds_box <-   
enc_dnds_ds_dn %>%
  mutate(enc_bin = cut(enc, breaks=c(floor(min(enc)), round(quantile(enc)[[2]],3), round(quantile(enc)[[3]],3), round(quantile(enc)[[4]],3), ceiling(max(enc))))) %>%
  ggplot(aes(branch, dnds, fill=enc_bin)) +
  geom_boxplot(outlier.colour = "#8C888B") +
  geom_hline(yintercept = median(enc_dnds_ds_dn$dnds)) +
  #geom_point() +
  stat_compare_means(label.sep = "\n") +
  scale_x_discrete(name = "") +
  scale_y_continuous(trans = "log10",
                     name = expression(omega)) +
  scale_fill_manual(values = branch_colours[1:4],
                    labels = c("21 - 50", "50 - 54.5", "54.5 - 57.4", "57.4 - 61"),
                    name = expression("ENC"[obs])) +
  theme(panel.background = element_blank(),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_line(size = 0.2, colour = "black"),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        axis.text.x = element_text(size = TXT_SIZE, colour = "black", face = "italic", angle = 45, hjust = 1),
        legend.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.key = element_blank()) #+
  facet_grid(~branch)

ggsave(plot=ggarrange(
         ggarrange(plot_corr_enc_gc3_all, plot_corr_enc_diff_gc3_all, common.legend = T),
         enc_dnds_box, nrow = 2),
         file="figures/corr_enc_gc3.pdf", device = "pdf",height = 6, width=8)


  
#enc_dnds_ds_dn <- 
enc_dnds_ds_dn %>%
  mutate(diff_bin = cut(diff_enc, breaks=c(floor(min(diff_enc)), round(quantile(diff_enc)[[2]],3), round(quantile(diff_enc)[[3]],3), round(quantile(diff_enc)[[4]],3), ceiling(max(diff_enc))))) %>%
  group_by(branch, diff_bin) %>%
  dplyr::summarise(mean_dnds=mean(dnds)) #%>%
  

for (i in seq(1,8,1)) {
print(levels(enc_dnds_ds_dn$branch)[i])
print(
pairwise.wilcox.test(enc_dnds_ds_dn[enc_dnds_ds_dn$branch==levels(enc_dnds_ds_dn$branch)[i],]$dnds, enc_dnds_ds_dn[enc_dnds_ds_dn$branch==levels(enc_dnds_ds_dn$branch)[i],]$diff_bin)
)
}

#ds
#plot_corr_enc_dS <- 
  enc_dnds_ds_dn %>%
  ggplot(aes(enc, log(dS))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10", 
                     name = expression(italic(d[S]))) +
  scale_x_continuous(name = "ENC_obs") +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.key = element_blank())

legend_enc_branch <- get_legend(plot_corr_enc_dS)
ggsave(plot=ggarrange(
  plot_corr_enc_gc3_all, plot_corr_enc_dS, enc_dnds_box, legend_enc_branch, 
  nrow = 2, ncol = 2,
  labels = c("a", "b", "c", "")),
         file="figures/figure_enc.pdf", device = "pdf",height = 12, width=12)

summary(lm(dS~branch +enc+ gc_3, data = enc_dnds_ds_dn))
summary(lm(dS~enc*branch+ gc_3*branch, data = enc_dnds_ds_dn))
summary(lm(dS~branch +enc*gc_3, data = enc_dnds_ds_dn))

for (i in seq(1,8,1)) {
print(levels(enc_dnds_ds_dn$branch)[i])
print(
summary(lm(dS~enc*gc_3, data = enc_dnds_ds_dn[enc_dnds_ds_dn$branch==levels(enc_dnds_ds_dn$branch)[i],]))
)
}

for (i in seq(1,8,1)) {
print(levels(enc_dnds_ds_dn$branch)[i])
print(
summary(lm(log(dS)~scale(enc)*scale(gc_3), data = enc_dnds_ds_dn[enc_dnds_ds_dn$branch==levels(enc_dnds_ds_dn$branch)[i],]))
)
}

scale(enc_dnds_ds_dn$enc)

hist(log(enc_dnds_ds_dn$dS), breaks = 100
     )

```




```{r iv_plot_ds_types_test}
str(enc_gc_type)

  for (i in seq(1,4,1)) {
print(levels(enc_gc_type$type)[i])
print(
summary(lm(log(ds)~scale(enc)*scale(gc_3), data = enc_gc_type[enc_gc_type$type==levels(enc_gc_type$type)[i],]))
)
}

  for (i in seq(1,8,1)) {
print(levels(enc_gc_type$branch)[i])
print(
summary(lm(log(ds)~type + scale(enc)*scale(gc_3), data = enc_gc_type[enc_gc_type$branch==levels(enc_gc_type$branch)[i],]))
)
}

summary(lm(log(ds)~scale(enc)*scale(gc_3) + type, data = enc_gc_type))



summary(lm(log(ds)~scale(enc)*scale(gc_3), data = enc_gc_type_gc_cons[enc_gc_type_gc_cons$branch=="L. sinapis", ]))

summary(lm(log(ds)~scale(enc)*scale(gc_3), data = enc_gc_type_gc_cons[enc_gc_type_gc_cons$branch=="P. sennae", ]))

###############
hist(log(enc_gc_type$ds), breaks = 100
     )

enc_gc_type_all <- 
enc_gc_type %>%
  filter(type %in% "all")

for (i in seq(1,8)) {
  b <- levels(enc_gc_type_all$branch)[i]
  branch_data <- enc_gc_type_all[enc_gc_type_all$branch==b,]

model_ds <- lm(log(ds)~scale(gc_3)*scale(diff_enc), data = branch_data)

if (i==1) {
  mod_list_1 <- list(assign(paste("model_ds",b, sep="_"), model_ds))
} else {
  mod_list_1[[length(mod_list_1)+1]] <- assign(paste("model_ds",b, sep="_"), model_ds)
}
}



#gives false p-value stars!!!!
#use to get F-stats but use tab_model for creating table
stargazer(mod_list_1, type = "html", ci=TRUE,digits = 3, out="result_files/ms_model_ds_all_fstat.doc", p=NULL, p.auto = FALSE, dep.var.caption = " ")

#citation("stargazer")
write(levels(enc_gc_type$branch), file = "result_files/ms_model_ds_all_fstat.doc", append = T)


tab_model(mod_list_1, file = "result_files/ms_model_ds_all.doc", digits = 3, show.p = FALSE, p.style = "numeric_stars", collapse.ci = TRUE, ci.hyphen = ",")

#plot_models(mod_list, m.labels = gsub("model_B_", "",names(mod_list_red)),p.shape = TRUE, axis.lim = c(-0.5, 0.5))


plot_models(mod_list_1, legend.title = "dS all substitutions", m.labels = levels(enc_gc_type$branch), colors = branch_colours, p.shape = TRUE, spacing = 0.7,
            axis.labels = c("gc_3"="GC3", "enc"=expression(ENC[diff]), "gc_3:enc"=expression(GC3:ENC[diff]))) +
  scale_y_continuous() +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
          axis.text.x = element_text(size = TXT_SIZE, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = TXT_SIZE, face="italic"))


#ds
enc_gc_type_gc_cons <- 
enc_gc_type %>%
  filter(type %in% "gc_cons")

for (i in seq(1,8)) {
  b <- levels(enc_gc_type_gc_cons$branch)[i]
  branch_data <- enc_gc_type_gc_cons[enc_gc_type_gc_cons$branch==b,]

model_ds <- lm(log(ds)~scale(gc_3)*scale(diff_enc), data = branch_data)

if (i==1) {
  mod_list_1 <- list(assign(paste("model_ds",b, sep="_"), model_ds))
} else {
  mod_list_1[[length(mod_list_1)+1]] <- assign(paste("model_ds",b, sep="_"), model_ds)
}
}



#gives false p-value stars!!!!
#use to get F-stats but use tab_model for creating table
stargazer(mod_list_1, type = "html", ci=TRUE,digits = 3, out="result_files/ms_model_ds_gc_cons_fstat.doc", p=NULL, p.auto = FALSE, dep.var.caption = " ")

#citation("stargazer")
write(levels(enc_gc_type$branch), file = "result_files/ms_model_ds_gc_cons_fstat.doc", append = T)


tab_model(mod_list_1, file = "result_files/ms_model_ds_gc_cons.doc", digits = 3, show.p = FALSE, p.style = "numeric_stars", collapse.ci = TRUE, ci.hyphen = ",")

#plot_models(mod_list, m.labels = gsub("model_B_", "",names(mod_list_red)),p.shape = TRUE, axis.lim = c(-0.5, 0.5))


plot_models(mod_list_1, legend.title = "dS GC-conservative substitutions", m.labels = levels(enc_gc_type$branch), colors = branch_colours, p.shape = TRUE, spacing = 0.7,
            axis.labels = c("gc_3"="GC3", "enc"=expression(ENC[diff]), "gc_3:enc"=expression(GC3:ENC[diff]))) +
  scale_y_continuous() +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
          axis.text.x = element_text(size = TXT_SIZE, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = TXT_SIZE, face="italic"))


ggsave("figures/model_ds_est.pdf", device = "pdf", height = 8, width = 8)


plot_models(mod_list_red, legend.title = "", m.labels = lineages, colors = branch_colours, p.shape = TRUE, spacing = 0.7,
            axis.labels = c("gc_3"="GC3", "enc"="ENC", "dS"=expression(italic(d[S])), "diff_enc"=expression(ENC[diff]))) +
  scale_y_continuous(limits=c(-0.5, 0.5)) +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = 10, colour = "black"),
          axis.text.x = element_text(size = 10, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = 10, face="italic"))

ggsave("figures/model_est_red.pdf", device = "pdf", height = 4, width = 6)

plot_models(mod_list_red, legend.title = "", m.labels = lineages, colors = branch_colours, p.shape = TRUE, spacing = 0.7,
            axis.labels = c("gc_3"="GC3", "enc"="ENC", "diff_enc"=expression(ENC[diff]))) +
  scale_y_continuous(limits=c(-0.5, 0.5)) +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = 10, colour = "black"),
          axis.text.x = element_text(size = 10, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = 10, face="italic"))

ggsave("figures/model_est_red_new.pdf", device = "pdf", height = 4, width = 6)

```
```{r model_dnds}

enc_gc_type_all <- 
enc_gc_type %>%
  filter(type %in% "all")

for (i in seq(1,8)) {
  b <- levels(enc_gc_type_all$branch)[i]
  branch_data <- enc_gc_type_all[enc_gc_type_all$branch==b,]

model_dnds <- lm(log(dnds)~scale(gc_3)*scale(diff_enc), data = branch_data)

if (i==1) {
  mod_list_1 <- list(assign(paste("model_dnds",b, sep="_"), model_dnds))
} else {
  mod_list_1[[length(mod_list_1)+1]] <- assign(paste("model_dnds",b, sep="_"), model_dnds)
}
}



#gives false p-value stars!!!!
#use to get F-stats but use tab_model for creating table
stargazer(mod_list_1, type = "html", ci=TRUE,digits = 3, out="result_files/ms_model_dnds_all_fstat.doc", p=NULL, p.auto = FALSE, dep.var.caption = " ")

#citation("stargazer")
write(levels(enc_gc_type$branch), file = "result_files/ms_model_dnds_all_fstat.doc", append = T)


tab_model(mod_list_1, file = "result_files/ms_model_dnds_all.doc", digits = 3, show.p = FALSE, p.style = "numeric_stars", collapse.ci = TRUE, ci.hyphen = ",")

#plot_models(mod_list, m.labels = gsub("model_B_", "",names(mod_list_red)),p.shape = TRUE, axis.lim = c(-0.5, 0.5))


plot_models(mod_list_1, legend.title = "dnds all substitutions", m.labels = levels(enc_gc_type$branch), colors = branch_colours, p.shape = TRUE, spacing = 0.7,
            axis.labels = c("gc_3"="GC3", "enc"=expression(ENC[diff]), "gc_3:enc"=expression(GC3:ENC[diff]))) +
  scale_y_continuous() +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
          axis.text.x = element_text(size = TXT_SIZE, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = TXT_SIZE, face="italic"))


#ds
enc_gc_type_gc_cons <- 
enc_gc_type %>%
  filter(type %in% "gc_cons")

for (i in seq(1,8)) {
  b <- levels(enc_gc_type_gc_cons$branch)[i]
  branch_data <- enc_gc_type_gc_cons[enc_gc_type_gc_cons$branch==b,]

model_dnds <- lm(log(dnds)~scale(gc_3) + scale(enc) + scale(diff_enc), data = branch_data)

if (i==1) {
  mod_list_2 <- list(assign(paste("model_dnds",b, sep="_"), model_dnds))
} else {
  mod_list_2[[length(mod_list_2)+1]] <- assign(paste("model_dnds",b, sep="_"), model_dnds)
}
}



#gives false p-value stars!!!!
#use to get F-stats but use tab_model for creating table
stargazer(mod_list_2, type = "html", ci=TRUE,digits = 3, out="result_files/ms_model_dnds_gc_cons_enc_fstat.doc", p=NULL, p.auto = FALSE, dep.var.caption = " ")

#citation("stargazer")
write(levels(enc_gc_type$branch), file = "result_files/ms_model_dnds_gc_cons_enc_fstat.doc", append = T)


tab_model(mod_list_2, file = "result_files/ms_model_dnds_gc_cons_enc.doc", digits = 3, show.p = FALSE, p.style = "numeric_stars", collapse.ci = TRUE, ci.hyphen = ",")

#plot_models(mod_list, m.labels = gsub("model_B_", "",names(mod_list_red)),p.shape = TRUE, axis.lim = c(-0.5, 0.5))

model_dnds <- 
plot_models(mod_list_1, legend.title="", title = "dnds GC-conservative substitutions", m.labels = levels(enc_gc_type$branch), colors = rev(branch_colours), p.shape = TRUE, spacing = 0.7,
            axis.labels = c("scale(gc_3)"="GC3", "scale(diff_enc)"="ENCdiff", "scale(gc_3):scale(diff_enc)"="GC3:ENCdiff")) +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
  scale_y_continuous(limits = c(-0.5, 0.5)) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
          axis.text.x = element_text(size = TXT_SIZE, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = TXT_SIZE, face="italic"))
          
model_dnds_2 <- 
plot_models(mod_list_2, legend.title="", title = expression(omega[ CG-cons]), m.labels = levels(enc_gc_type$branch), colors = rev(branch_colours), p.shape = TRUE, spacing = 0.7,
            axis.labels = c("scale(gc_3)"="GC3", "scale(enc)"="ENC", "scale(diff_enc)"="ENCdiff")) +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
  scale_y_continuous(limits = c(-0.5, 0.5)) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
          axis.text.x = element_text(size = TXT_SIZE, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = TXT_SIZE, face="italic"))
model_dnds
model_dnds_2

legend_pvalue <- get_legend(model_dnds_2)[[1]]$'99_f3326d1149b10be16e34bce2d43a0d04'
model_dnds_2 + theme(legend.position = "none") + legend_pvalue

plot(enc_gc_type$enc, (enc_gc_type$exp_enc-enc_gc_type$enc))
plot(enc_gc_type$enc, enc_gc_type$diff_enc)

hist(enc_gc_type$diff_enc)



for (i in seq(1,8)) {
  b <- levels(enc_gc_type_gc_cons$branch)[i]
  branch_data <- enc_gc_type_gc_cons[enc_gc_type_gc_cons$branch==b,]

model_dnds <- lm(dnds~scale(gc_3) + scale(diff_enc) +, data = branch_data)

if (i==1) {
  mod_list_1 <- list(assign(paste("model_dnds",b, sep="_"), model_dnds))
} else {
  mod_list_1[[length(mod_list_1)+1]] <- assign(paste("model_dnds",b, sep="_"), model_dnds)
}
}

plot_models(mod_list_1, legend.title="", title = "dnds GC-conservative substitutions", m.labels = levels(enc_gc_type$branch), colors = rev(branch_colours), p.shape = TRUE, spacing = 0.7) +
  geom_hline(yintercept = 0, colour = "black", size = 0.2) +
  scale_y_continuous(limits = c(-0.1, 0.1)) +
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size=0.2),
          axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
          axis.text.x = element_text(size = TXT_SIZE, colour = "black"),
          axis.ticks.x = element_line(size = 0.2, colour = "black"),
          axis.ticks.y = element_blank(),
          legend.text = element_text(size = TXT_SIZE, face="italic"))


```


```{r iv_plot_types_plots}

dn_enc <- 
 enc_gc_type %>%
   #filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(enc, log(dn))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[N])*")")) +
  scale_x_continuous(name = expression(ENC[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)
 
dn_enc_diff <- 
 enc_gc_type %>%
   #filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(diff_enc, log(dn))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[N])*")")) +
  scale_x_continuous(name = expression(ENC[diff])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ggsave(plot=ggarrange(dn_enc, dn_enc_diff, common.legend = T, legend = "bottom", nrow = 2),
       filename = "figures_ms/dn_enc_types.pdf",
       device = "pdf",
       height = 12, width = 12
       )

ds_enc <- 
  enc_gc_type %>%
   #filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(enc, log(ds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[S])*")")) +
  scale_x_continuous(name = expression(ENC[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ds_enc_diff <- 
  enc_gc_type %>%
   #filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(diff_enc, log(ds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[S])*")")) +
  scale_x_continuous(name = expression(ENC[diff])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ggsave(plot=ggarrange(ds_enc, ds_enc_diff, common.legend = T, legend = "bottom", nrow = 2),
       filename = "figures_ms/ds_enc_types.pdf",
       device = "pdf",
       height = 12, width = 12
       )

dnds_enc <- 
  enc_gc_type %>%
   #filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(enc, log(dnds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*omega*")")) +
  scale_x_continuous(name = expression(ENC[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

dnds_enc_diff <- 
  enc_gc_type %>%
   #filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(diff_enc, log(dnds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*omega*")")) +
  scale_x_continuous(name = expression(ENC[diff])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ggsave(plot=ggarrange(dnds_enc, dnds_enc_diff, common.legend = T, legend = "bottom", nrow = 2),
       filename = "figures_ms/dnds_enc_types.pdf",
       device = "pdf",
       height = 12, width = 12
       )



########
#suppl all types enc and enc diff
dnds_enc_diff_2 <- 
  enc_gc_type %>%
   filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(diff_enc, log(dnds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*omega*")")) +
  scale_x_continuous(name = expression(ENC[diff])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ds_enc_diff_2 <- 
  enc_gc_type %>%
   filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(diff_enc, log(ds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[S])*")")) +
  scale_x_continuous(name = expression(ENC[diff])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

dn_enc_diff_2 <- 
  enc_gc_type %>%
   filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(diff_enc, log(dn))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[N])*")")) +
  scale_x_continuous(name = expression(ENC[diff])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ggarrange(dnds_enc_diff_2 + rremove("xlab") + rremove("x.text"), dn_enc_diff_2 + rremove("xlab") + rremove("x.text"), ds_enc_diff_2, 
          common.legend = T,
          ncol = 1,
          heights = c(1,1,1,2)
          )


dnds_enc_2 <- 
  enc_gc_type %>%
   filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(enc, log(dnds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y.npc = "bottom", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*omega*")")) +
  scale_x_continuous(name = expression(ENC[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ds_enc_2 <- 
  enc_gc_type %>%
   filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(enc, log(ds))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y.npc = "bottom", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[S])*")")) +
  scale_x_continuous(name = expression(ENC[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

dn_enc_2 <- 
  enc_gc_type %>%
   filter(type %in% c("all", "gc_cons")) %>%
     ggplot(aes(enc, log(dn))) +
  geom_point(aes(colour=branch), alpha=0.05) +
  geom_smooth(aes(colour = branch), method = "lm", se=F) +
  stat_cor(aes(colour=branch), size=2, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y.npc = "bottom", show.legend=FALSE) +
  scale_y_continuous(name = expression("log("*italic(d[N])*")")) +
  scale_x_continuous(name = expression(ENC[obs])) +
  scale_colour_manual(values = branch_colours) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        legend.key = element_blank(),
        legend.title = element_blank()) +
    facet_grid(~type)

ggarrange(dnds_enc_2 + rremove("xlab") + rremove("x.text"), dn_enc_2 + rremove("xlab") + rremove("x.text"), ds_enc_2, 
          common.legend = T,
          ncol = 1,
          heights = c(1,1,1,2)
          )



corr_enc_gc3_gc_cons <- 
   enc_gc_type %>%
   filter(type %in% c("gc_cons")) %>%
  ggplot(aes(gc_3*100, enc)) +
  geom_point(aes(colour = branch), alpha=0.05) +
  #geom_smooth(aes(colour = branch), method = "lm") +
  geom_smooth(aes(colour = branch), method = "loess") +
  geom_smooth(aes(gc_3*100, exp_enc), colour = "#494446", linetype = "dashed", method = "loess") +
  #stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y = -0.40) +
  scale_y_continuous(name = expression("ENC"[obs])) +
  scale_x_continuous(name = "GC-content in third position (%)") +
  scale_color_manual(values = branch_colours) +
  #scale_fill_gradient(high = "#541608", low = "#d6b8ab") +
  #facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        axis.title = element_text(size = TXT_SIZE, colour = "black"),
        axis.line = element_line(size = 0.2, colour = "black"),
        axis.text = element_text(size = TXT_SIZE, colour = "black"),
        legend.text = element_text(size = TXT_SIZE, colour = "black"))


ggarrange(dnds_enc + rremove("xlab") + rremove("x.text"), dn_enc + rremove("xlab") + rremove("x.text"), ds_enc, corr_enc_gc3_gc_cons, model_dnds,
          common.legend = T,
          ncol = 2,
          heights = c(1,1,1,2)
          )

p1=dnds_enc_2 + rremove("xlab") + rremove("x.text") + theme(legend.position = "none")
p2=dn_enc_2 + rremove("xlab") + rremove("x.text") + theme(legend.position = "none")
p3=ds_enc_2 + theme(legend.position = "none")
p4=model_dnds_2 + theme(legend.position = "none", plot.margin = unit(c(5.5, 0, 5.5, 5.5), "pt")) + legend_pvalue + theme(plot.margin = unit(c(5.5, 0, 5.5, 5.5), "pt"))
p5=corr_enc_gc3_gc_cons + theme(legend.position = "none")
p6=get_legend(dnds_enc_2)

ggsave(
  (p1 + p5 + p2 + (p4 + p4b) + p3 + p6) +
  plot_layout(ncol=2) + plot_annotation(tag_levels = list(c("a", "d", "b", "e","", "c"))),
       filename = "figures_ms/enc_plot_ms.pdf", device = "pdf", height = 12, width = 12)


```


```{r iv_enc_dataset_suppl}

enc_gc_type[enc_gc_type$dataset=="pos_sel" & !is.na(enc_gc_type$dataset), "dataset"]="neutral"

enc_dataset <- 
enc_gc_type %>%
   filter(type %in% c("gc_cons")) %>%
  filter(!is.na(enc)) %>%
  ggplot(aes(dataset, enc)) +
    geom_boxplot(aes(fill=dataset),outlier.fill = "grey", outlier.size = 0.3) + 
     #scale_fill_manual(values=c("#7D3232", "#D4AF7D", "white"),
       #                name="Gene set", 
       #                labels=c("cons"="Conserved", "neutral"="Aligned")) +
     xlab(label = "") +
     ylab(label = expression(ENC[obs])) +
     scale_fill_manual(values=c("#7D3232", "#D4AF7D"),
                       name="Gene set", 
                       labels=c("cons"="Conserved", "neutral"="Rest of aligned")) +
  stat_compare_means(method = "wilcox",label.sep = "\n") +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
        axis.text.x = element_blank(),
        axis.title = element_text(size = TXT_SIZE), 
        legend.key = element_blank(), 
        legend.text = element_text(size = TXT_SIZE, colour = "black"))

enc_diff_dataset <- 
enc_gc_type %>%
   filter(type %in% c("gc_cons")) %>%
  filter(!is.na(diff_enc)) %>%
  ggplot(aes(dataset, diff_enc)) +
    geom_boxplot(aes(fill=dataset), outlier.fill = "grey", outlier.size = 0.3) + 
     xlab(label = "") +
     ylab(label = expression(ENC[diff])) +
     scale_fill_manual(values=c("#7D3232", "#D4AF7D"),
                       name="Gene set", 
                       labels=c("cons"="Conserved", "neutral"="Rest of aligned")) +
  stat_compare_means(method = "wilcox",label.sep = "\n") +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = TXT_SIZE, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = TXT_SIZE, colour = "black"),
        axis.text.x = element_blank(),
        axis.title = element_text(size = TXT_SIZE), 
        legend.key = element_blank(), 
        legend.text = element_text(size = TXT_SIZE, colour = "black"))


ggsave(plot=ggarrange(enc_dataset, enc_diff_dataset, ncol = 2), 
       filename = "figures/enc_dataset_box_Suppl.pdf",
       device = "pdf", height = 6, width = 12)



```


```{r iv_plot_dnds, echo=FALSE}
enc_dnds_ds_dn <- read.table(file = "../../07_codon_usage/07_03_codon_usage_analysis_r/data/enc_dnds_ds_dn.table", sep = " ")


#reassign
enc_dnds_ds_dn$branch <- as.factor(enc_dnds_ds_dn$branch)
levels(enc_dnds_ds_dn$branch) <- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. accius", "L. sinapis", "P. machaon", "P. sennae")

#reorder
enc_dnds_ds_dn$branch <- factor(enc_dnds_ds_dn$branch, levels=c("B. mori","P. machaon", "L. accius", "L. sinapis", "P. sennae", "C. cecrops", "D. plexippus", "H. melpomene"))

ds_color <- "#b38064"
dn_color <- "blue"
#enc and ds
plot_corr_enc_dS <- ggplot(enc_dnds_ds_dn, aes(enc, dS)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10", 
                     name = expression("log("*italic(d[S])*")")) +
  scale_x_continuous(name = "ENC_obs") +
  scale_fill_gradient(high = "white", low = ds_color) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_dS, file="figures/corr_enc_dS.pdf", device = "pdf",height = 6, width=8)

plot_corr_enc_diff_dS <- ggplot(enc_dnds_ds_dn, aes(diff_enc, dS)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10", 
                     name = expression("log("*italic(d[S])*")")) +
  scale_x_continuous(name = "ENC_diff") +
  scale_fill_gradient(high = "white", low = ds_color) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_diff_dS, file="figures/corr_enc_diff_dS.pdf", device = "pdf",height = 6, width=8)

#dN
plot_corr_enc_dN <- ggplot(enc_dnds_ds_dn, aes(enc, dN)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10", 
                     name = expression("log("*italic(d[N])*")")) +
  scale_x_continuous(name = "ENC_obs") +
  scale_fill_gradient(high = "white", low = dn_color) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_dN, file="figures/corr_enc_dN.pdf", device = "pdf",height = 6, width=8)

plot_corr_enc_diff_dN <- ggplot(enc_dnds_ds_dn, aes(diff_enc, dN)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black") +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10", 
                     name = expression("log("*italic(d[N])*")")) +
  scale_x_continuous(name = "ENC_diff") +
  scale_fill_gradient(high = "white", low = dn_color) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_diff_dN, file="figures/corr_enc_diff_dN.pdf", device = "pdf",height = 6, width=8)

dnds_colour <- "#778868"
#plot_corr_enc_w <- 
  ggplot(enc_dnds_ds_dn, aes(enc, dnds)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black", size = 0.4) +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left") +
  scale_y_continuous(trans="log10", 
                     name = expression("log("*omega*")")) +
  scale_x_continuous(name = expression("ENC"[obs])) +
  scale_fill_gradient(high = "white", low = dnds_colour) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_w, file="figures/corr_enc_w.pdf", device = "pdf",height = 6, width=8)
#f4fae1, #8a9183, #a1a899
plot_corr_enc_w_inv <- ggplot(enc_dnds_ds_dn, aes(dnds, enc)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black", size = 0.4) +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "centre", label.y = 22) +
  scale_x_continuous(trans="log10", 
                     name = expression("log("*omega*")")) +
  scale_y_continuous(name = expression("ENC"[obs])) +
  scale_fill_gradient(high = "#2a361f", low = "#abb89c") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_enc_w_inv, file="figures/corr_enc_w_inv.pdf", device = "pdf",height = 6, width=8)

#dnds and diff_enc
plot_corr_diff_enc_w <- ggplot(enc_dnds_ds_dn, aes(dnds, diff_enc)) +
  geom_bin2d() +
  #geom_smooth(method="loess", colour = "dark grey") +
  geom_smooth(method = "lm", colour="black", size = 0.4) +
  stat_cor(size=2.5, method = "spearman", cor.coef.name = "rho", label.x.npc = "left", label.y = -0.40) +
  scale_x_continuous(trans="log10", 
                     name = expression("log("*omega*")")) +
  scale_y_continuous(name = expression("ENC"[diff])) +
  scale_fill_gradient(high = "#2a361f", low = "#abb89c") +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
        panel.border = element_rect(size = 0.2, alpha(colour = "white",alpha = 0)),
        strip.background = element_blank(),
        strip.text = element_text(size = 10, face = "italic", colour = "black"), 
        axis.title = element_text(size = 10, colour = "black"),
        axis.line = element_blank(),
        axis.text = element_text(size = 10, colour = "black"),
        legend.text = element_text(size = 10, colour = "black"),
        legend.position = c(0.8, 0.15))

ggsave(plot=plot_corr_diff_enc_w, file="figures/corr_diff_enc_w.pdf", device = "pdf",height = 6, width=8)
```


```{r iv_plot_gc_again, echo=FALSE}
#plot enc and exp_enc
ggplot(enc_gc_long) + 
    geom_point(aes(enc_gc_long$gc_3, enc_gc_long$enc),
               fill = "grey",
               size = 1,
               shape = 21, 
               stroke = 0.05) +
    geom_line(aes(enc_gc_long$gc_3, enc_gc_long$exp_enc, colour = enc_gc_long$branch),
              size = 1.5) +
    scale_color_viridis(discrete = T,
                        guide_legend(title = "Expected enc per branch")) +
    facet_wrap(~enc_gc_long$branch, 2, 4) +
    xlab(label = "GC-content in third codon position") +
    ylab(label = "Effective number of codons (enc)") +
    theme(panel.background = element_blank(), 
          strip.background = element_blank(),
          panel.border = element_rect(size = 1, fill = alpha("white", 0)),
          strip.text = element_text(size = 10, face = "italic"), 
          axis.text = element_text(size = 8, colour = "black"), 
          legend.text = element_text(size = 10, face = "italic"),
          legend.key = element_blank()
          )


#one plot enc and exp_enc and lm obs enc
ggplot(enc_gc_long) + 
    geom_point(aes(gc_3, enc, fill = branch),
               size = 1,
               shape = 21, 
               stroke = 0.05) +
    geom_smooth(mapping = aes(gc_3, exp_enc), 
              colour ="#CDCCCE",
              size = 1.5, 
              linetype="dashed") +
    geom_smooth(aes(enc_gc_long$gc_3, enc_gc_long$enc, colour = enc_gc_long$branch), method = "loess", fill = alpha("white", 0)) +
    scale_color_viridis(discrete = T,
                        guide_legend(title = "Branch")) +
    scale_fill_viridis(alpha = 0.3, 
                       discrete = T,
                      guide = "none") +
      #facet_wrap(~enc_gc_long$branch, 2, 4) +
    xlab(label = "GC-content in third codon position") +
    ylab(label = "Effective number of codons (enc)") +
    theme(panel.background = element_blank(), 
          strip.background = element_blank(),
          panel.border = element_rect(size = 1, fill = alpha("white", 0)),
          strip.text = element_text(size = 10, face = "italic"), 
          axis.text = element_text(size = 8, colour = "black"), 
          legend.text = element_text(size = 10, face = "italic"),
          legend.key = element_blank()
          )

ggplot(enc_gc_long) + 
    geom_point(aes(gc_3, enc, fill = branch),
               size = 1,
               shape = 21, 
               stroke = 0.05,
               colour = "white") +
    geom_smooth(mapping = aes(gc_3, exp_enc), 
              colour ="#494446",
              size = 1.5, 
              linetype="dashed") +
    geom_smooth(aes(enc_gc_long$gc_3, enc_gc_long$enc, colour = enc_gc_long$branch), method = "loess", fill = alpha("white", 0)) +
    scale_color_viridis(discrete = T,
                        guide_legend(title = "Branch")) +
    scale_fill_viridis(alpha = 0.3, 
                       discrete = T,
                      guide = "none") +
      #facet_wrap(~enc_gc_long$branch, 2, 4) +
    xlab(label = "GC-content in third codon position") +
    ylab(label = "Effective number of codons (enc)") +
    theme(panel.background = element_blank(), 
          strip.background = element_blank(),
          panel.border = element_rect(size = 1, fill = alpha("white", 0)),
          strip.text = element_text(size = 10, face = "italic"), 
          axis.text = element_text(size = 8, colour = "black"), 
          legend.text = element_text(size = 10, face = "italic"),
          legend.key = element_blank()
          )

ggplot(enc_gc_long) + 
    geom_point(aes(gc_3, enc, fill = branch),
               size = 1,
               shape = 21, 
               stroke = 0.05,
               colour = "white") +
    geom_smooth(mapping = aes(gc_3, exp_enc), 
              colour ="#494446",
              size = 1.5, 
              linetype="dashed") +
    geom_smooth(aes(enc_gc_long$gc_3, enc_gc_long$enc, colour = enc_gc_long$branch), method = "loess", fill = alpha("white", 0)) +
    scale_color_carto_d(palette = "Safe",
                        guide_legend(title = "Branch")) +
    scale_fill_carto_d(palette = "Safe",
                      guide = "none") +
      #facet_wrap(~enc_gc_long$branch, 2, 4) +
    xlab(label = "GC-content in third codon position") +
    ylab(label = "Effective number of codons (enc)") +
    theme(panel.background = element_blank(), 
          strip.background = element_blank(),
          panel.border = element_rect(size = 1, fill = alpha("white", 0)),
          strip.text = element_text(size = 10, face = "italic"), 
          axis.text = element_text(size = 8, colour = "black"), 
          legend.text = element_text(size = 10, face = "italic"),
          legend.key = element_blank()
          )

branch_colours <- nord(n = 8, palette = "red_mountain")

branch_colours_alpha <- nord(n = 8, palette = "red_mountain", alpha = 0.1)

ggplot(enc_gc_long) + 
    geom_point(aes(gc_3, enc, fill = branch),
               size = 1,
               shape = 21, 
               stroke = 0.05,
               colour = "white") +
    geom_rect(aes(xmin=min(enc_gc_long$gc_3), xmax=max(enc_gc_long$gc_3), ymin=min(enc_gc_long$enc), ymax = max(enc_gc_long$enc)+1), fill="white", alpha=0.01) +
    geom_smooth(mapping = aes(gc_3, exp_enc), 
              colour ="#494446",
              size = 1.5, 
              linetype="dashed") +
    geom_smooth(aes(enc_gc_long$gc_3, enc_gc_long$enc, colour = enc_gc_long$branch), method = "loess", se = F) +
    scale_color_manual(values = branch_colours,
                        guide_legend(title = "Branch")) +
    scale_fill_manual(values = branch_colours,
                      guide = "none") +
      #facet_wrap(~enc_gc_long$branch, 2, 4) +
    xlab(label = "GC-content in third codon position") +
    ylab(label = "Effective number of codons (enc)") +
    theme(panel.background = element_blank(), 
          strip.background = element_blank(),
          panel.border = element_rect(size = 1, fill = alpha("white", 0)),
          strip.text = element_text(size = 10, face = "italic"), 
          axis.text = element_text(size = 8, colour = "black"), 
          legend.text = element_text(size = 10, face = "italic"),
          legend.key = element_blank()
          )

ggsave("figures/enc_gc_red_mount.pdf", device = "pdf", height = 4, width = 6)
#one plot enc and exp_enc and lm obs ecn

ggplot(enc_gc_long) + 
    # geom_point(aes(enc_gc_long$gc_3, enc_gc_long$enc),
    #            fill = "grey",
    #            size = 1,
    #            shape = 21, 
    #            stroke = 0.05) +
    geom_smooth(aes(enc_gc_long$gc_3, enc_gc_long$exp_enc), colour = "#494446", size=1.5, linetype="dashed") +
    geom_smooth(aes(enc_gc_long$gc_3, enc_gc_long$enc, colour = enc_gc_long$branch),fill = alpha("white", 0), method = "loess") +
    scale_color_viridis(discrete = T,
                        guide_legend(title = "Branch")) +
#    facet_wrap(~enc_gc_long$branch, 2, 4) +
    xlab(label = "GC-content in third codon position") +
    ylab(label = "Effective number of codons (enc)") +
    theme(panel.background = element_blank(), 
          strip.background = element_blank(),
          panel.border = element_rect(size = 1, fill = alpha("white", 0)),
          strip.text = element_text(size = 10, face = "italic"), 
          axis.text = element_text(size = 8, colour = "black"), 
          legend.text = element_text(size = 10, face = "italic"),
          legend.key = element_blank()
          )

#density plot
ggplot(enc_gc_long) + geom_density(aes(enc_gc_long$diff_enc, group = enc_gc_long$branch, fill = enc_gc_long$branch)) + 
     scale_fill_viridis(discrete = T, 
                        guide_legend(title = "Branch"), 
                        alpha = 0.5) +
     xlab(label = "Difference between observed and expected enc") +
     theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           legend.text = element_text(face = "italic"))

#violin plot
ggplot(enc_gc_long) + 
    geom_violin(aes(enc_gc_long$branch, enc_gc_long$diff_enc, fill = enc_gc_long$branch)) +     scale_fill_viridis(discrete = T, 
                        guide = "none") +
     xlab(label = "") +
     ylab(label = "ENCdiff") +
     theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           axis.text.x = element_text(face = "italic"))

#boxplot
ggplot(enc_gc_long) + 
    geom_boxplot(aes(enc_gc_long$branch, enc_gc_long$diff_enc, fill = enc_gc_long$branch)) +     scale_fill_manual(values = branch_colours) +
     xlab(label = "") +
     ylab(label = "ENCdiff") +
     theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           axis.text.x = element_text(face = "italic"))


ggplot(enc_gc_long, aes(gc_3, diff_enc)) + 
  geom_point(aes(fill = branch), shape = 21, size = 2) + 
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "spearman") +
  scale_fill_manual(values = branch_colours) +
  xlab(label = "") +
  scale_y_continuous(name = expression("ENC"[diff])) +
  theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           axis.text.x = element_text(face = "italic"))

#boxplot enc obs
ggplot(enc_gc_long) + 
    geom_boxplot(aes(enc_gc_long$branch, enc_gc_long$enc, fill = enc_gc_long$branch)) +     scale_fill_manual(values = branch_colours) +
     xlab(label = "") +
     ylab(label = "ENC") +
     theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           axis.text.x = element_text(face = "italic"))


enc_gc_mean <- left_join(aggregate(enc~branch, mean, data=enc_gc_long), aggregate(gc_3~branch, mean, data=enc_gc_long))

enc_gc_mean <- left_join(enc_gc_mean, aggregate(diff_enc~branch, mean, data=enc_gc_long))
enc_gc_mean <- left_join(enc_gc_mean, aggregate(enc~branch, mean, data=enc_gc_long))
enc_gc_mean <- left_join(enc_gc_mean, aggregate(enc~branch, sd, data=enc_gc_long))


#against gc
#dotplot enc_diff
box_enc_diff_vs_gc3 <- ggplot(enc_gc_mean, aes(gc_3, diff_enc)) + 
  geom_point(aes(gc_3, diff_enc, fill = enc_gc_mean$branch), shape = 21, size = 5) + 
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "spearman") +
  scale_fill_manual(values = branch_colours) +
  xlab(label = "GC-content at third position") +
  scale_y_continuous(name = expression("ENC"[diff]),
                     limits = c(0.03, 0.07)) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())

ggsave(box_enc_diff_vs_gc3, filename = "figures/box_enc_diff_vs_gc3.pdf", device = "pdf")

#mean enc obs vs gc
box_enc_vs_gc3 <- ggplot(enc_gc_mean, aes(gc_3, enc)) + 
  geom_point(aes(fill = branch), shape = 21, size = 5) + 
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "spearman") +
  scale_fill_manual(values = branch_colours) +
  xlab(label = "GC-content at third position") +
  scale_y_continuous(name = expression("ENC"[obs])) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())

ggsave(box_enc_vs_gc3, filename = "figures/box_enc_vs_gc3.pdf", device = "pdf")

box_enc_vs_gc3_not_cecrops <- ggplot(enc_gc_mean, aes(gc_3, enc)) + 
  geom_point(aes(fill = branch), shape = 21, size = 5) + 
  geom_smooth(method = "lm", se = F, size = 0.4) +
  geom_smooth(data = subset(enc_gc_mean, branch!="C. cecrops"), aes(gc_3, enc), method = "lm", se = F, size = 0.4, linetype = "dashed") +
  stat_cor(data = subset(enc_gc_mean, branch!="C. cecrops"), aes(gc_3, enc), method = "spearman") +
  scale_fill_manual(values = branch_colours) +
  xlab(label = "GC-content at third position") +
  scale_y_continuous(name = expression("ENC"[obs])) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())

ggsave(box_enc_vs_gc3_not_cecrops, filename = "figures/box_enc_vs_gc3_not_cecrops.pdf", device = "pdf")

aggregate(gc_3~branch, sd, data=enc_gc_long)

#mean gc3 vs sd gc3
gc3mean_gc3sd <- ggplot(enc_gc_mean, aes(gc_3*100, aggregate(gc_3~branch, sd, data=enc_gc_long)$gc_3*100)) + 
  geom_point(aes(fill = branch), shape = 21, size = 5) + 
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "pearson") +
  scale_fill_manual(values = branch_colours) +
  scale_x_continuous(name = "Mean GC-content at third position (%)", 
                     limits = c(35,60)) +
  scale_y_continuous(name = "Standard deviation GC3", 
                     limits = c(10,20)) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())

ggsave(gc3mean_gc3sd, filename = "figures/gc3mean_gc3sd.pdf", device = "pdf")

enc_gc_long$at_3 <- 1-enc_gc_long$gc_3
at_mean <- aggregate(at_3~branch, mean, data=enc_gc_long)
#mean at3 vs sd at3
at3mean_at3sd <- ggplot(at_mean, aes(at_3*100, aggregate(at_3~branch, sd, data=enc_gc_long)$at_3*100)) + 
  geom_point(aes(fill = branch), shape = 21, size = 5) + 
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "pearson") +
  scale_fill_manual(values = branch_colours) +
  scale_x_continuous(name = "Mean AT-content at third position (%)", 
                     limits = c(35,65)) +
  scale_y_continuous(name = "Standard deviation AT3", 
                     limits = c(10,20)) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())

ggsave(at3mean_at3sd, filename = "figures/at3mean_at3sd.pdf", device = "pdf")


#gc3 vs sd top and bottom 20%
gc3mean_gc3sd_low <- ggplot(enc_gc_mean) + 
  geom_point(aes(gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long)$gc_3,fill = branch), shape = 21, size = 5) + 
  geom_smooth(aes(gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long)$gc_3), method = "lm", se = F) +
  stat_cor(aes(gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long)$gc_3), method = "spearman", label.x = 0.35, label.y.npc = "top") +
  geom_point(aes(aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3, fill = branch), shape = 21, size = 5) + 
  geom_smooth(aes(aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3), method = "lm", se = F) +
  stat_cor(aes(aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3), method = "spearman", label.x = 0.3, label.y = 0.01) +
  geom_point(aes(aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3, fill = branch), shape = 21, size = 5) + 
  geom_smooth(aes(aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3), method = "lm", se = F) +
  stat_cor(aes(aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3, aggregate(gc_3~branch, sd, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3), method = "spearman", label.x = 0.6, label.y.npc = "centre") +
  scale_fill_manual(values = branch_colours) +
  scale_x_continuous(name = "Mean GC-content in third codon position", 
                     limits = c(0.15,0.8)) +
  scale_y_continuous(name = "Standard deviation GC3", 
                     limits = c(0,0.2)) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())

ggsave(gc3mean_gc3sd_low, filename = "figures/gc3mean_gc3sd_top_bottom.pdf", device = "pdf", height = 6, width = 8)
#gc3 top and bottom vs mean 
gc3mean_gc3low_high <- ggplot(enc_gc_mean) + 
  geom_point(aes(gc_3, aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3, fill = branch), shape = 21, size = 5) + 
  geom_smooth(aes(gc_3, aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3), method = "lm", se = F) +
  stat_cor(aes(gc_3, aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3<quantile(enc_gc_long$gc_3, probs=0.2, na.rm=T),])$gc_3), method = "spearman", label.x.npc = "centre", label.y.npc = "bottom") +
  geom_point(aes(gc_3, aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3, fill = branch), shape = 21, size = 5) + 
  geom_smooth(aes(gc_3, aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3), method = "lm", se = F) +
  stat_cor(aes(gc_3, aggregate(gc_3~branch, mean, data=enc_gc_long[enc_gc_long$gc_3>quantile(enc_gc_long$gc_3, probs=0.81, na.rm=T),])$gc_3), method = "spearman", label.x.npc = "centre") +
  scale_fill_manual(values = branch_colours) +
  xlab(label = "GC-content at third position (mean)") +
  scale_y_continuous(name = "GC-content at third position (mean) 20% extreme bin", 
                     limits = c(0,1)) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())

#enc mean vs sd
ggplot(enc_gc_mean, aes(enc, aggregate(enc~branch, sd, data=enc_gc_long)$enc)) + 
  geom_point(aes(fill = branch), shape = 21, size = 5) + 
  geom_smooth(method = "lm", se = F) +
  stat_cor(method = "pearson") +
  scale_fill_manual(values = branch_colours) +
  scale_x_continuous(name = expression("Mean ENC"[obs])) +
  scale_y_continuous(name = expression("Standard deviation ENC"[obs])) +
  theme(panel.background = element_blank(),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        legend.text = element_text(face = "italic"), 
           legend.title = element_blank())


ggplot(enc_gc_long) + 
    geom_violin(aes(enc_gc_long$branch, enc_gc_long$diff_enc, fill=enc_gc_long$dataset, group=interaction(enc_gc_long$branch, enc_gc_long$dataset))) +     scale_fill_viridis(discrete = T) +
     xlab(label = "") +
     ylab(label = "Diff enc") +
     theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           axis.text.x = element_text(face = "italic"))

#boxplot enc diff greyscale
ggplot(subset(enc_gc_long, dataset!="pos_sel")) + 
    geom_boxplot(aes(dataset, diff_enc, fill=dataset), outlier.fill = "grey", outlier.size = 0.3) + 
     scale_fill_manual(values=c("#777777", "#CCCCCC"),
                       name="Dataset", 
                       labels=c("cons"="Conserved", "neutral"="Aligned")) +
     xlab(label = "") +
     ylab(label = "Diff enc") +
  stat_compare_means(aes(dataset, diff_enc), 
                     method = "wilcox",label.sep = "\n", label.y = 0.42) +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = 10, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = 10, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title = element_text(size = 10), 
        legend.key = element_blank(), 
        legend.text = element_text(size = 10, colour = "black"))

ggsave("figures/boxplot_enc_diff_dataset.pdf", device = "pdf", height = 6, width = 8)

boxplot_enc_diff_dataset_brown <- ggplot(subset(enc_gc_long, dataset!="pos_sel")) + 
    geom_boxplot(aes(dataset, diff_enc, fill=dataset), outlier.fill = "grey", outlier.size = 0.3) + 
     scale_fill_manual(values=c("#7D3232", "#D4AF7D"),
                       name="Gene set", 
                       labels=c("cons"="Conserved", "neutral"="Aligned")) +
     xlab(label = "") +
     ylab(label = expression("ENC"[diff])) +
  stat_compare_means(aes(dataset, diff_enc), 
                     method = "wilcox",label.sep = "\n", label.y = 0.42) +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = 10, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = 10, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title = element_text(size = 10), 
        legend.key = element_blank(), 
        legend.text = element_text(size = 10, colour = "black"))

ggsave("figures/boxplot_enc_diff_dataset_brown.pdf", device = "pdf", height = 6, width = 8)

#obs enc
boxplot_enc_dataset_brown <- ggplot(subset(enc_gc_long, dataset!="pos_sel")) + 
    geom_boxplot(aes(dataset, enc, fill=dataset), outlier.fill = "grey", outlier.size = 0.3) + 
     scale_fill_manual(values=c("#7D3232", "#D4AF7D"),
                       name="Gene set", 
                       labels=c("neutral"="Aligned", "cons"="Conserved")) +
     scale_x_discrete(name = "", 
                      ) +
     scale_y_continuous(name = expression("ENC"[obs]),
                        limits = c(20,70)) +
  stat_compare_means(aes(dataset, enc), 
                     method = "wilcox",label.sep = "\n") +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = 10, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = 10, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title = element_text(size = 10), 
        legend.key = element_blank(), 
        legend.text = element_text(size = 10, colour = "black"))

ggsave("figures/boxplot_enc_dataset_brown.pdf", device = "pdf", height = 6, width = 8)


enc_gc_long$dataset <- as.factor(enc_gc_long$dataset)
my_comp <- list(c("cons", "neutral"), c("neutral", "pos_sel"))

#all datasets
ggplot(enc_gc_long) + 
    geom_boxplot(aes(dataset, diff_enc, fill=dataset), outlier.fill = "grey", outlier.size = 0.3) + 
     #scale_fill_manual(values=c("#7D3232", "#D4AF7D", "white"),
       #                name="Gene set", 
       #                labels=c("cons"="Conserved", "neutral"="Aligned", "pos_sel"="Pos sel")) +
     xlab(label = "") +
     ylab(label = "Diff enc") +
  stat_compare_means(aes(dataset, diff_enc), comparisons = my_comp, method = "wilcox",label.sep = "\n") +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = 10, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = 10, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title = element_text(size = 10), 
        legend.key = element_blank(), 
        legend.text = element_text(size = 10, colour = "black"))

ggsave("figures/boxplot_enc_diff_dataset_brown_all.pdf", device = "pdf", height = 6, width = 8)

#incl CDS
enc_CDS <- read.table("data/enc_CDS.table")
colnames(enc_CDS) <- c("GeneID", "enc", "branch")
enc_CDS$dataset <- "CDS" 
#reassign
enc_CDS$branch <- as.factor(enc_CDS$branch)
levels(enc_CDS$branch) <- c("B. mori", "C. cecrops", "D. plexippus", "H. melpomene", "L. sinapis",  "L. accius", "P. machaon", "P. sennae")

#reorder
enc_CDS$branch <- factor(enc_CDS$branch, levels=c("B. mori","P. machaon", "L. accius", "L. sinapis", "P. sennae", "C. cecrops", "D. plexippus", "H. melpomene"))


enc_CDS_aligned_cons <- rbind(enc_CDS, enc_gc_long[,c("GeneID", "enc", "branch", "dataset")])
enc_CDS_aligned_cons[enc_CDS_aligned_cons$dataset=="pos_sel","dataset"] <- "Aligned"
enc_CDS_aligned_cons[enc_CDS_aligned_cons$dataset=="neutral","dataset"] <- "Aligned"
enc_CDS_aligned_cons[enc_CDS_aligned_cons$dataset=="cons","dataset"] <- "Conserved"

enc_CDS_aligned_cons$dataset <- as.factor(enc_CDS_aligned_cons$dataset)
#reorder
enc_CDS_aligned_cons$dataset <- factor(enc_CDS_aligned_cons$dataset, levels=c("CDS", "Aligned", "Conserved"))

my_comp <- list(c("CDS", "Aligned"), c("Conserved", "Aligned"))

boxplot_enc_dataset_brown_CDS <- ggplot(enc_CDS_aligned_cons) + 
    geom_boxplot(aes(dataset, enc, fill=dataset), outlier.fill = "grey", outlier.size = 0.3) + 
    scale_fill_manual(values=c("red", "#7D3232", "#D4AF7D"),
                      name="Gene set") + 
     scale_x_discrete(name = "", 
                      ) +
     scale_y_continuous(name = expression("ENC"[obs]),
                        limits = c(20,70)) +
  stat_compare_means(aes(dataset, enc), 
                     method = "wilcox",label.sep = "\n") +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = 10, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = 10, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title = element_text(size = 10), 
        legend.key = element_blank(), 
        legend.text = element_text(size = 10, colour = "black"))

ggsave(boxplot_enc_dataset_brown_CDS, "figures/boxplot_enc_dataset_brown_CDS.pdf", device = "pdf", height = 6, width = 8)


enc_CDS_aligned <- ggplot(enc_CDS_aligned_mean) + 
    geom_point(aes(dataset, enc, fill=dataset), outlier.fill = "grey", outlier.size = 0.3) + 
    scale_fill_manual(values=c("dark grey","#7D3232", "#D4AF7D"),
                      name="Gene set") +
     scale_x_discrete(name = "", 
                      ) +
     scale_y_continuous(name = expression("ENC"[obs]),
                        limits = c(20,70)) +
  #stat_compare_means(aes(dataset, enc), 
  #                   method = "wilcox",label.sep = "\n") +
  facet_wrap(~branch, nrow = 2, ncol = 4) +
  theme(panel.background = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.border = element_rect(size = 0.2, fill = alpha("white", 0)),
        strip.background = element_rect(colour = "black", fill=alpha("white", 0)),
        strip.text = element_text(size = 10, colour = "black", face = "italic"),
        axis.line.y.left = element_line(size = 0.2),
        axis.line.x.bottom = element_line(size = 0.2),
        axis.text.y = element_text(size = 10, colour = "black"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title = element_text(size = 10), 
        legend.key = element_blank(), 
        legend.text = element_text(size = 10, colour = "black"))

ggsave(boxplot_enc_dataset_brown_CDS, "figures/boxplot_enc_dataset_brown_CDS.pdf", device = "pdf", height = 6, width = 8)


ggplot(enc_gc_long) + 
    geom_boxplot(aes(enc_gc_long$dataset, enc_gc_long$diff_enc, fill=enc_gc_long$branch)) +
     scale_fill_manual(values=branch_colours) +
     xlab(label = "") +
     ylab(label = "Diff enc") +
     theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           axis.text.x = element_text(face = "italic"))

wilcox_pvalue <- cbind(levels(enc_gc_long$branch), numeric(length = length(1:8)))
for (i in (1:8)) {
  b=levels(enc_gc_long$branch)[i]
 wilcox_pvalue[i, 2] <- wilcox.test(subset(enc_gc_long, branch==b & dataset=="cons")$diff_enc, subset(enc_gc_long, branch==b & dataset=="neutral")$diff_enc)$p.value
}

capture.output(pairwise.wilcox.test(enc_gc_long$diff_enc, enc_gc_long$dataset), file = "result_files/enc_dataset_wilcox.txt")
capture.output(pairwise.wilcox.test(subset(enc_gc_long, branch=="L. sinapis")$diff_enc, subset(enc_gc_long, branch=="L. sinapis")$dataset, p.adjust.method = "none"), file = "result_files/enc_dataset_wilcox.txt", append = T)

write("Wilcoxon rank sum test with continuity correction: \nTesting difference in diff_enc between conserved and aligned dataset.\n", file = "result_files/enc_dataset_wilcox.txt", append = T)
write.table(wilcox_pvalue, file = "result_files/enc_dataset_wilcox.txt", append = T)

aggregate(diff_enc~dataset + branch, data=enc_gc_long, mean)

#enc obs
ggsave(ggarrange(plot_corr_enc_gc3, box_enc_vs_gc3_not_cecrops, plot_corr_enc_w_inv, boxplot_enc_dataset_brown, ncol = 2, nrow = 2), filename = "figures/multi_enc.pdf", device = "pdf", height = 10, width = 12)


#enc diff
ggsave(ggarrange(plot_corr_enc_diff_gc3, box_enc_diff_vs_gc3, plot_corr_diff_enc_w, boxplot_enc_diff_dataset_brown, ncol = 2, nrow = 2), filename = "figures/multi_enc_diff.pdf", device = "pdf", height = 10, width = 12)


#enc diff histo
ggplot(enc_gc_long) + 
  geom_histogram(aes(diff_enc, fill = branch), binwidth = 0.01) + 
  geom_vline(xintercept = quantile(enc_gc_long$diff_enc)) +
  scale_fill_manual(values = branch_colours, 
                    guide = "none") +
  scale_x_continuous(name = expression("ENC"[diff])) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           strip.text = element_text(face = "italic"))


ggplot(enc_gc_long) + 
  geom_histogram(aes(enc, fill = branch), binwidth = 1) + 
  geom_vline(xintercept = quantile(enc_gc_long$enc)) +
  scale_fill_manual(values = branch_colours, 
                    guide = "none") +
  scale_x_continuous(name = expression("ENC"[obs])) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           strip.text = element_text(face = "italic"))

ggplot(enc_CDS) + 
  geom_histogram(aes(enc, fill = branch), binwidth = 1) + 
  geom_vline(xintercept = quantile(enc_CDS$enc)) +
  scale_fill_manual(values = branch_colours, 
                    guide = "none") +
  scale_x_continuous(name = expression("ENC"[obs])) +
  facet_wrap(~branch) +
  theme(panel.background = element_blank(),
           panel.border = element_rect(size = 1, fill = alpha("white", 0)),
           strip.text = element_text(face = "italic"))


#check Dros
enc_dros <- read.table("data/enc_CDS_prel.txt_Dros")
aggregate(V2~V3, median, data=enc_dros)
aggregate(V2~V3, mean, data=enc_dros)

```

